
<!DOCTYPE html>

<html>
<head>
<meta charset="utf-8"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<title>1.3.4: Learning and Evolution — Modern Foundations of Neuroscience</title>
<!-- Loaded before other Sphinx assets -->
<link href="../../../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet"/>
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet"/>
<link href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css" rel="stylesheet"/>
<link as="font" crossorigin="" href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2" rel="preload" type="font/woff2"/>
<link as="font" crossorigin="" href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2" rel="preload" type="font/woff2"/>
<link href="../../../_static/pygments.css" rel="stylesheet" type="text/css">
<link href="../../../_static/styles/sphinx-book-theme.css" rel="stylesheet" type="text/css">
<link href="../../../_static/togglebutton.css" rel="stylesheet" type="text/css">
<link href="../../../_static/copybutton.css" rel="stylesheet" type="text/css">
<link href="../../../_static/mystnb.css" rel="stylesheet" type="text/css">
<link href="../../../_static/sphinx-thebe.css" rel="stylesheet" type="text/css"/>
<link href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" rel="stylesheet" type="text/css"/>
<link href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" rel="stylesheet" type="text/css"/>
<!-- Pre-loaded scripts that we'll load fully later -->
<link as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf" rel="preload"/>
<script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
<script src="../../../_static/jquery.js"></script>
<script src="../../../_static/underscore.js"></script>
<script src="../../../_static/doctools.js"></script>
<script src="../../../_static/togglebutton.js"></script>
<script src="../../../_static/clipboard.min.js"></script>
<script src="../../../_static/copybutton.js"></script>
<script src="../../../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
<script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
<script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
<script async="async" src="../../../_static/sphinx-thebe.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
<script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.1/dist/embed-amd.js"></script>
<script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
<link href="../../../genindex.html" rel="index" title="Index">
<link href="../../../search.html" rel="search" title="Search"/>
<link href="../../P1C4_LifeIsRL/student/P1C4_Title.html" rel="next" title="Chapter 1.4 Life is Reinforcement Learning"/>
<link href="P1C3_Sequence3.html" rel="prev" title="1.3.3: Competition and Interaction Transform Optimization into Game Theory"/>
<meta content="width=device-width, initial-scale=1" name="viewport"/>
<meta content="None" name="docsearch:language"/>
<!-- Google Analytics -->
</link></link></link></link></link></link></head>
<body data-offset="60" data-spy="scroll" data-target="#bd-toc-nav">
<!-- Checkboxes to toggle the left sidebar -->
<input aria-label="Toggle navigation sidebar" class="sidebar-toggle" id="__navigation" name="__navigation" type="checkbox"/>
<label class="overlay overlay-navbar" for="__navigation">
<div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input aria-label="Toggle in-page Table of Contents" class="sidebar-toggle" id="__page-toc" name="__page-toc" type="checkbox"/>
<label class="overlay overlay-pagetoc" for="__page-toc">
<div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>
<div class="container-fluid" id="banner"></div>
<div class="container-xl">
<div class="row">
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
<div class="bd-sidebar__content">
<div class="bd-sidebar__top"><div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../../../index.html">
<h1 class="site-logo" id="site-title">Modern Foundations of Neuroscience</h1>
</a>
</div><form action="../../../search.html" class="bd-search d-flex align-items-center" method="get">
<i class="icon fas fa-search"></i>
<input aria-label="Search this book..." autocomplete="off" class="form-control" id="search-input" name="q" placeholder="Search this book..." type="search"/>
</form><nav aria-label="Main" class="bd-links" id="bd-docs-nav">
<div class="bd-toc-item active">
<ul class="nav bd-sidenav bd-sidenav__home-link">
<li class="toctree-l1">
<a class="reference internal" href="../../intro.html">
                    Introduction
                </a>
</li>
</ul>
<p class="caption">
<span class="caption-text">
  Behaviour, Environments and Optimization, Evolution and Learning
 </span>
</p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../P1C1_BehaviourAsPolicy/student/P1C1_Title.html">
   Behaviour As Policy (P1C1)
  </a>
<input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox">
<label for="toctree-checkbox-1">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2">
<a class="reference internal" href="../../P1C1_BehaviourAsPolicy/student/P1C1_Sequence1.html">
<strong>
      1.1.1: Gridworld Introduction
     </strong>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../P1C1_BehaviourAsPolicy/student/P1C1_Sequence2.html">
<strong>
      1.1.2: Thinking of What Receptors and Muscles Do as a Policy
     </strong>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../P1C1_BehaviourAsPolicy/student/P1C1_Sequence3.html">
<strong>
      1.1.3: Abstraction and Reasoning with Policies
     </strong>
</a>
</li>
</ul>
</input></li>
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../P1C2_OptimizationAndEnvironment/student/P1C2_Title.html">
   Optimization And Environment (P1C2)
  </a>
<input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox">
<label for="toctree-checkbox-2">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2">
<a class="reference internal" href="../../P1C2_OptimizationAndEnvironment/student/P1C2_Sequence1.html">
<strong>
      1.2.1: Optimization Makes you Better
     </strong>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../P1C2_OptimizationAndEnvironment/student/P1C2_Sequence2.html">
<strong>
      1.2.2: What’s Good Depends on the Niche
     </strong>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../P1C2_OptimizationAndEnvironment/student/P1C2_Sequence3.html">
<strong>
      1.2.3: Normative Thinking
     </strong>
</a>
</li>
</ul>
</input></li>
<li class="toctree-l1 current active has-children">
<a class="reference internal" href="P1C3_Title.html">
   Real Evolution (P1C3)
  </a>
<input checked="" class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox">
<label for="toctree-checkbox-3">
<i class="fas fa-chevron-down">
</i>
</label>
<ul class="current">
<li class="toctree-l2">
<a class="reference internal" href="P1C3_Sequence1.html">
<strong>
      1.3.1: Populations and Inheritance
     </strong>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="P1C3_Sequence2.html">
<strong>
      1.3.2: Why Sex
     </strong>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="P1C3_Sequence3.html">
<strong>
      1.3.3: Competition and Interaction Transform Optimization into Game Theory
     </strong>
</a>
</li>
<li class="toctree-l2 current active">
<a class="current reference internal" href="#">
<strong>
      1.3.4: Learning and Evolution
     </strong>
</a>
</li>
</ul>
</input></li>
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../P1C4_LifeIsRL/student/P1C4_Title.html">
   Life Is RL (P1C4)
  </a>
<input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
<label for="toctree-checkbox-4">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2">
<a class="reference internal" href="../../P1C4_LifeIsRL/student/P1C4_Sequence1.html">
<strong>
      1.4.1: Thinking of Life as a Markov Decision Process
     </strong>
</a>
</li>
</ul>
</li>
</ul>
<p class="caption">
<span class="caption-text">
  Supervised Learning
 </span>
</p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children">
<a class="reference internal" href="../../P2C1_Optimization/student/P2C1_Title.html">
   Optimization (P2C1)
  </a>
<input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
<label for="toctree-checkbox-5">
<i class="fas fa-chevron-down">
</i>
</label>
<ul>
<li class="toctree-l2">
<a class="reference internal" href="../../P2C1_Optimization/student/P2C1_Sequence1.html">
<strong>
      2.1.1: Optimization in One Dimension: Developing Intuition
     </strong>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../P2C1_Optimization/student/P2C1_Sequence2.html">
<strong>
      2.1.2: Optimization in Two Dimensions: Estimating Gradients
     </strong>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../P2C1_Optimization/student/P2C1_Sequence3.html">
<strong>
      2.1.3: Optimization in Higher Dimensions: The Need for Speed
     </strong>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../P2C1_Optimization/student/P2C1_Sequence4.html">
<strong>
      2.1.4: Learning Behaviour as a Form of High Dimensional Optimization
     </strong>
</a>
</li>
<li class="toctree-l2">
<a class="reference internal" href="../../P2C1_Optimization/student/P2C1_Sequence5.html">
<strong>
      2.1.5: Learning Behaviour Quickly
     </strong>
</a>
</li>
</ul>
</li>
</ul>
</div>
</nav></div>
<div class="bd-sidebar__bottom">
<!-- To handle the deprecated key -->
<div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>
</div>
</div>
<div id="rtd-footer-container"></div>
</div>
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
<div class="header-article row sticky-top noprint">
<div class="col py-1 d-flex header-article-main">
<div class="header-article__left">
<label class="headerbtn" data-placement="right" data-toggle="tooltip" for="__navigation" title="Toggle navigation">
<span class="headerbtn__icon-container">
<i class="fas fa-bars"></i>
</span>
</label>
</div>
<div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
<button aria-label="Launch interactive content" class="headerbtn menu-dropdown__trigger">
<i class="fas fa-rocket"></i>
</button>
<div class="menu-dropdown__content">
<ul>
</ul>
</div>
</div>
<button class="headerbtn" data-placement="bottom" data-toggle="tooltip" onclick="toggleFullScreen()" title="Fullscreen mode">
<span class="headerbtn__icon-container">
<i class="fas fa-expand"></i>
</span>
</button>
<div class="menu-dropdown menu-dropdown-repository-buttons">
<button aria-label="Source repositories" class="headerbtn menu-dropdown__trigger">
<i class="fab fa-github"></i>
</button>
<div class="menu-dropdown__content">
<ul>
<li>
<a class="headerbtn" data-placement="left" data-toggle="tooltip" href="https://github.com/dcownden/PerennialProblemsOfLifeWithABrain" title="Source repository">
<span class="headerbtn__icon-container">
<i class="fab fa-github"></i>
</span>
<span class="headerbtn__text-container">repository</span>
</a>
</li>
<li>
<a class="headerbtn" data-placement="left" data-toggle="tooltip" href="https://github.com/dcownden/PerennialProblemsOfLifeWithABrain/issues/new?title=Issue%20on%20page%20%2Fsequences/P1C3_RealEvolution/student/P1C3_Sequence4.html&amp;body=Your%20issue%20content%20here." title="Open an issue">
<span class="headerbtn__icon-container">
<i class="fas fa-lightbulb"></i>
</span>
<span class="headerbtn__text-container">open issue</span>
</a>
</li>
</ul>
</div>
</div>
<div class="menu-dropdown menu-dropdown-download-buttons">
<button aria-label="Download this page" class="headerbtn menu-dropdown__trigger">
<i class="fas fa-download"></i>
</button>
<div class="menu-dropdown__content">
<ul>
<li>
<a class="headerbtn" data-placement="left" data-toggle="tooltip" href="../../../_sources/sequences/P1C3_RealEvolution/student/P1C3_Sequence4.ipynb" title="Download source file">
<span class="headerbtn__icon-container">
<i class="fas fa-file"></i>
</span>
<span class="headerbtn__text-container">.ipynb</span>
</a>
</li>
<li>
<button class="headerbtn" data-placement="left" data-toggle="tooltip" onclick="printPdf(this)" title="Print to PDF">
<span class="headerbtn__icon-container">
<i class="fas fa-file-pdf"></i>
</span>
<span class="headerbtn__text-container">.pdf</span>
</button>
</li>
</ul>
</div>
</div>
<label class="headerbtn headerbtn-page-toc" for="__page-toc">
<span class="headerbtn__icon-container">
<i class="fas fa-list"></i>
</span>
</label>
</div>
</div>
<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
<div class="tocsection onthispage pt-5 pb-3">
<i class="fas fa-list"></i> Contents
    </div>
<nav aria-label="Page" id="bd-toc-nav">
<ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry">
<a class="reference internal nav-link" href="#">
<strong>
    1.3.4: Learning and Evolution
   </strong>
</a>
<ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#objective">
     Objective:
    </a>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry">
<a class="reference internal nav-link" href="#setup">
   Setup
  </a>
</li>
<li class="toc-h1 nav-item toc-entry">
<a class="reference internal nav-link" href="#learning-as-a-hedge-against-uncertainty">
   1.3.4.1: Learning as a Hedge Against Uncertainty
  </a>
<ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#coding-exercise-implement-minimal-trial-and-error-learning">
     Coding Exercise: Implement Minimal Trial and Error Learning
    </a>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry">
<a class="reference internal nav-link" href="#some-fitness-score-functions-are-hard-for-evolution-to-optimize">
   1.3.4.2: Some Fitness Score Functions are Hard for Evolution to Optimize
  </a>
<ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#coding-exercise-implement-the-new-fitness-score-function">
     Coding Exercise: Implement the New Fitness Score Function.
    </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#math-exercise-probabilities-of-evolutionary-events">
     Math Exercise: Probabilities of Evolutionary Events.
    </a>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry">
<a class="reference internal nav-link" href="#learning-can-accelerate-adaptive-response">
   1.3.4.3 Learning Can Accelerate Adaptive Response
  </a>
<ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#todo-pciture-evolution-and-learning-as-inner-and-outer-loops-of-one-optimization-process">
<strong>
      TODO
     </strong>
     Pciture: Evolution and Learning as Inner and Outer Loops of One Optimization Process.
    </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#todo-bio-box-facilitated-variation">
<strong>
      TODO
     </strong>
     Bio Box: Facilitated Variation
    </a>
</li>
</ul>
</li>
</ul>
</nav>
</div>
</div>
<div class="article row">
<div class="col pl-md-3 pl-lg-5 content-container">
<!-- Table of contents that is only displayed when printing the page -->
<div class="onlyprint" id="jb-print-docs-body">
<h1>1.3.4: Learning and Evolution</h1>
<!-- Table of contents -->
<div id="print-main-content">
<div id="jb-print-toc">
<div>
<h2> Contents </h2>
</div>
<nav aria-label="Page">
<ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry">
<a class="reference internal nav-link" href="#">
<strong>
    1.3.4: Learning and Evolution
   </strong>
</a>
<ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#objective">
     Objective:
    </a>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry">
<a class="reference internal nav-link" href="#setup">
   Setup
  </a>
</li>
<li class="toc-h1 nav-item toc-entry">
<a class="reference internal nav-link" href="#learning-as-a-hedge-against-uncertainty">
   1.3.4.1: Learning as a Hedge Against Uncertainty
  </a>
<ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#coding-exercise-implement-minimal-trial-and-error-learning">
     Coding Exercise: Implement Minimal Trial and Error Learning
    </a>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry">
<a class="reference internal nav-link" href="#some-fitness-score-functions-are-hard-for-evolution-to-optimize">
   1.3.4.2: Some Fitness Score Functions are Hard for Evolution to Optimize
  </a>
<ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#coding-exercise-implement-the-new-fitness-score-function">
     Coding Exercise: Implement the New Fitness Score Function.
    </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#math-exercise-probabilities-of-evolutionary-events">
     Math Exercise: Probabilities of Evolutionary Events.
    </a>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry">
<a class="reference internal nav-link" href="#learning-can-accelerate-adaptive-response">
   1.3.4.3 Learning Can Accelerate Adaptive Response
  </a>
<ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#todo-pciture-evolution-and-learning-as-inner-and-outer-loops-of-one-optimization-process">
<strong>
      TODO
     </strong>
     Pciture: Evolution and Learning as Inner and Outer Loops of One Optimization Process.
    </a>
</li>
<li class="toc-h2 nav-item toc-entry">
<a class="reference internal nav-link" href="#todo-bio-box-facilitated-variation">
<strong>
      TODO
     </strong>
     Bio Box: Facilitated Variation
    </a>
</li>
</ul>
</li>
</ul>
</nav>
</div>
</div>
</div>
<main id="main-content" role="main">
<div>
<p><a href="https://colab.research.google.com/github/dcownden/PerennialProblemsOfLifeWithABrain/blob/main/sequences/P1C3_RealEvolution/student/P1C3_Sequence4.ipynb" target="_blank"><img alt="Open In Colab" src="https://colab.research.google.com/assets/colab-badge.svg"/></a>   <a href="https://kaggle.com/kernels/welcome?src=https://raw.githubusercontent.com/dcownden/PerennialProblemsOfLifeWithABrain/main/sequences/P1C3_RealEvolution/student/P1C3_Sequence4.ipynb" target="_blank"><img alt="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"/></a></p>
<p>The following is part of a test for an upcoming text book on computational neuroscience from an optimization and learning perspective. The book will start with evolution because ultimately, all aspects of the brain are shaped by evolution and, as we will see, evolution can also be seen as an optimization algorithm. We are sharing it now to get feedback on what works and what does not and the developments we should do.</p>
<hr class="docutils"/>
<div class="section" id="learning-and-evolution">
<h1><strong>1.3.4: Learning and Evolution</strong><a class="headerlink" href="#learning-and-evolution" title="Permalink to this headline">¶</a></h1>
<p>In the last sequence we saw how the evolutionary process itself creates ever changing environments and consequently selective pressure for rapid adaptation. In the sequence before we saw how sex and recombination can drastically accelerate the evolutionary process, i.e. adaptation over generational time. In this sequence we turn to adaptation ‘within lifetime’. Within lifetime adaptation is sometimes called phenotypic plasticity, with learned behaviour as a kind of pinnacle of phenotypic plasticity. For long lived organisms with complex and flexible behaviours, ongoing adaptive plasticity learning is critically important.</p>
<div class="section" id="objective">
<h2>Objective:<a class="headerlink" href="#objective" title="Permalink to this headline">¶</a></h2>
<p>Appreciate the evolutionary advantage of within lifetime adaptation, both as a hedge against environmental variability and uncertainty, and as an accelerator of response to selection.</p>
<p>In this sequence we explore the following questions through simulations and mathematical analyses:</p>
<ul class="simple">
<li><p>How and under what conditions does within lifetime adaptation (learning) serve as a hedge against a changing environment?</p></li>
<li><p>How and under what conditions does within lifetime adaptation (learning) increase the rate of response to changing selective pressures.</p></li>
<li><p>Why do we expect more within lifetime learning in long-lived animals with behavioural flexibility?</p></li>
</ul>
</div>
</div>
<div class="section" id="setup">
<h1>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h1>
<p>Run the following cell to setup and install the various dependencies and helper functions for this .</p>
<p><strong>Dependencies, Imports and Setup.</strong> You don’t need to worry about how this code works – but you do need to <strong>run the cell</strong></p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># @markdown **Dependencies, Imports and Setup.** You don't need to worry about how this code works – but you do need to **run the cell**</span>
<span class="o">!</span>apt<span class="w"> </span>install<span class="w"> </span>libgraphviz-dev<span class="w"> </span>&gt;<span class="w"> </span>/dev/null<span class="w"> </span><span class="m">2</span>&gt;<span class="w"> </span>/dev/null<span class="w"> </span>#colab
<span class="o">!</span>pip<span class="w"> </span>install<span class="w"> </span>ipympl<span class="w"> </span>pygraphviz<span class="w"> </span>vibecheck<span class="w"> </span>datatops<span class="w"> </span>jupyterquiz<span class="w"> </span>&gt;<span class="w"> </span>/dev/null<span class="w"> </span><span class="m">2</span>&gt;<span class="w"> </span>/dev/null<span class="w"> </span>#google.colab

<span class="kn">import</span> <span class="nn">requests</span>
<span class="kn">from</span> <span class="nn">requests.exceptions</span> <span class="kn">import</span> <span class="n">RequestException</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib.patches</span> <span class="k">as</span> <span class="nn">mpatches</span>
<span class="kn">from</span> <span class="nn">matplotlib.animation</span> <span class="kn">import</span> <span class="n">FuncAnimation</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">pygraphviz</span> <span class="k">as</span> <span class="nn">pgv</span>
<span class="kn">import</span> <span class="nn">ipywidgets</span> <span class="k">as</span> <span class="nn">widgets</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">copy</span>
<span class="kn">import</span> <span class="nn">torch</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">torch.nn</span> <span class="k">as</span> <span class="nn">nn</span>
<span class="kn">import</span> <span class="nn">torch.optim</span> <span class="k">as</span> <span class="nn">optim</span>
<span class="kn">import</span> <span class="nn">torch.nn.functional</span> <span class="k">as</span> <span class="nn">F</span>
<span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">BytesIO</span>
<span class="kn">from</span> <span class="nn">enum</span> <span class="kn">import</span> <span class="n">Enum</span>
<span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">cdist</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">norm</span>
<span class="kn">from</span> <span class="nn">scipy.optimize</span> <span class="kn">import</span> <span class="n">minimize</span>
<span class="kn">from</span> <span class="nn">sklearn.linear_model</span> <span class="kn">import</span> <span class="n">LinearRegression</span>
<span class="kn">from</span> <span class="nn">tabulate</span> <span class="kn">import</span> <span class="n">tabulate</span>
<span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">display</span><span class="p">,</span> <span class="n">clear_output</span><span class="p">,</span> <span class="n">Markdown</span><span class="p">,</span> <span class="n">HTML</span><span class="p">,</span> <span class="n">Image</span>
<span class="kn">from</span> <span class="nn">jupyterquiz</span> <span class="kn">import</span> <span class="n">display_quiz</span>
<span class="kn">from</span> <span class="nn">vibecheck</span> <span class="kn">import</span> <span class="n">DatatopsContentReviewContainer</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Dict</span>
<span class="kn">from</span> <span class="nn">tqdm.notebook</span> <span class="kn">import</span> <span class="n">tqdm</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">"ignore"</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">UserWarning</span><span class="p">,</span> <span class="n">module</span><span class="o">=</span><span class="s2">"matplotlib"</span><span class="p">)</span>
<span class="c1"># random seed settings and</span>
<span class="c1"># getting torch to use gpu if it's there</span>


<span class="k">def</span> <span class="nf">set_seed</span><span class="p">(</span><span class="n">seed</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">seed_torch</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">  </span><span class="sd">"""</span>
<span class="sd">  Function that controls randomness. NumPy and random modules must be imported.</span>

<span class="sd">  Args:</span>
<span class="sd">    seed : Integer</span>
<span class="sd">      A non-negative integer that defines the random state. Default is `None`.</span>
<span class="sd">    seed_torch : Boolean</span>
<span class="sd">      If `True` sets the random seed for pytorch tensors, so pytorch module</span>
<span class="sd">      must be imported. Default is `True`.</span>

<span class="sd">  Returns:</span>
<span class="sd">    Nothing.</span>
<span class="sd">  """</span>
  <span class="k">if</span> <span class="n">seed</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">seed</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">2</span> <span class="o">**</span> <span class="mi">32</span><span class="p">)</span>
  <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
  <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">seed_torch</span><span class="p">:</span>
    <span class="n">torch</span><span class="o">.</span><span class="n">manual_seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">manual_seed_all</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">manual_seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">torch</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">cudnn</span><span class="o">.</span><span class="n">benchmark</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">torch</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">cudnn</span><span class="o">.</span><span class="n">deterministic</span> <span class="o">=</span> <span class="kc">True</span>

  <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Random seed </span><span class="si">{</span><span class="n">seed</span><span class="si">}</span><span class="s1"> has been set.'</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">seed_worker</span><span class="p">(</span><span class="n">worker_id</span><span class="p">):</span>
<span class="w">  </span><span class="sd">"""</span>
<span class="sd">  DataLoader will reseed workers following randomness in</span>
<span class="sd">  multi-process data loading algorithm.</span>

<span class="sd">  Args:</span>
<span class="sd">    worker_id: integer</span>
<span class="sd">      ID of subprocess to seed. 0 means that</span>
<span class="sd">      the data will be loaded in the main process</span>
<span class="sd">      Refer: https://pytorch.org/docs/stable/data.html#data-loading-randomness for more details</span>

<span class="sd">  Returns:</span>
<span class="sd">    Nothing</span>
<span class="sd">  """</span>
  <span class="n">worker_seed</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">initial_seed</span><span class="p">()</span> <span class="o">%</span> <span class="mi">2</span><span class="o">**</span><span class="mi">32</span>
  <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">worker_seed</span><span class="p">)</span>
  <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">worker_seed</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">set_device</span><span class="p">():</span>
<span class="w">  </span><span class="sd">"""</span>
<span class="sd">  Set the device. CUDA if available, CPU otherwise</span>

<span class="sd">  Args:</span>
<span class="sd">    None</span>

<span class="sd">  Returns:</span>
<span class="sd">    Nothing</span>
<span class="sd">  """</span>
  <span class="n">device</span> <span class="o">=</span> <span class="s2">"cuda"</span> <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span> <span class="k">else</span> <span class="s2">"cpu"</span>
  <span class="k">if</span> <span class="n">device</span> <span class="o">!=</span> <span class="s2">"cuda"</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">"This notebook isn't using and doesn't need a GPU. Good."</span><span class="p">)</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">"GPU is enabled in this notebook but not needed."</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">"If possible, in the menu under `Runtime` -&gt; "</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">"`Change runtime type.`  select `CPU`"</span><span class="p">)</span>

  <span class="k">return</span> <span class="n">device</span>


<span class="n">SEED</span> <span class="o">=</span> <span class="mi">2021</span>
<span class="n">set_seed</span><span class="p">(</span><span class="n">seed</span><span class="o">=</span><span class="n">SEED</span><span class="p">)</span>
<span class="n">DEVICE</span> <span class="o">=</span> <span class="n">set_device</span><span class="p">()</span>


<span class="k">def</span> <span class="nf">printmd</span><span class="p">(</span><span class="n">string</span><span class="p">):</span>
  <span class="n">display</span><span class="p">(</span><span class="n">Markdown</span><span class="p">(</span><span class="n">string</span><span class="p">))</span>


<span class="c1"># the different utility .py files used in this notebook</span>
<span class="n">filenames</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># just run the code straight out of the response, no local copies needed!</span>
<span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">:</span>
  <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">'https://raw.githubusercontent.com/dcownden/PerennialProblemsOfLifeWithABrain/main/utils/</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">'</span>
  <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
  <span class="c1"># Check that we got a valid response</span>
  <span class="k">if</span> <span class="n">response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span><span class="p">:</span>
    <span class="n">code</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">content</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
    <span class="n">exec</span><span class="p">(</span><span class="n">code</span><span class="p">)</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Failed to download </span><span class="si">{</span><span class="n">url</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>

<span class="c1"># environment contingent imports</span>
<span class="k">try</span><span class="p">:</span>
  <span class="nb">print</span><span class="p">(</span><span class="s1">'Running in colab'</span><span class="p">)</span>
  <span class="kn">from</span> <span class="nn">google.colab</span> <span class="kn">import</span> <span class="n">output</span>
  <span class="n">output</span><span class="o">.</span><span class="n">enable_custom_widget_manager</span><span class="p">()</span>
  <span class="kn">from</span> <span class="nn">google.colab</span> <span class="kn">import</span> <span class="n">data_table</span>
  <span class="n">data_table</span><span class="o">.</span><span class="n">disable_dataframe_formatter</span><span class="p">()</span>
  <span class="c1">#from google.colab import output as colab_output</span>
  <span class="c1">#colab_output.enable_custom_widget_manager()</span>
  <span class="n">IN_COLAB</span> <span class="o">=</span> <span class="kc">True</span>
<span class="k">except</span><span class="p">:</span>
  <span class="n">IN_COLAB</span> <span class="o">=</span> <span class="kc">False</span>
  <span class="nb">print</span><span class="p">(</span><span class="s1">'Not running in colab'</span><span class="p">)</span>

<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = 'retina'
<span class="o">%</span><span class="k">matplotlib</span> widget
<span class="n">plt</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s2">"https://raw.githubusercontent.com/dcownden/PerennialProblemsOfLifeWithABrain/main/pplb.mplstyle"</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ioff</span><span class="p">()</span> <span class="c1">#need to use plt.show() or display explicitly</span>
<span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s1">'matplotlib.font_manager'</span><span class="p">)</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">ERROR</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">remove_ip_clutter</span><span class="p">(</span><span class="n">fig</span><span class="p">):</span>
  <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">header_visible</span> <span class="o">=</span> <span class="kc">False</span>
  <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">toolbar_visible</span> <span class="o">=</span> <span class="kc">False</span>
  <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">resizable</span> <span class="o">=</span> <span class="kc">False</span>
  <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">footer_visible</span> <span class="o">=</span> <span class="kc">False</span>
  <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>


<span class="k">def</span> <span class="nf">content_review</span><span class="p">(</span><span class="n">notebook_section</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
  <span class="k">return</span> <span class="n">DatatopsContentReviewContainer</span><span class="p">(</span>
    <span class="s2">""</span><span class="p">,</span>  <span class="c1"># No text prompt</span>
    <span class="n">notebook_section</span><span class="p">,</span>
    <span class="p">{</span>
      <span class="s2">"url"</span><span class="p">:</span> <span class="s2">"https://pmyvdlilci.execute-api.us-east-1.amazonaws.com/klab"</span><span class="p">,</span>
      <span class="s2">"name"</span><span class="p">:</span> <span class="s2">"neuro_book"</span><span class="p">,</span>
      <span class="s2">"user_key"</span><span class="p">:</span> <span class="s2">"xuk960xj"</span><span class="p">,</span>
    <span class="p">},</span>
  <span class="p">)</span><span class="o">.</span><span class="n">render</span><span class="p">()</span>
<span class="n">feedback_prefix</span> <span class="o">=</span> <span class="s2">"P1C3_S4"</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Random seed 2021 has been set.
This notebook isn't using and doesn't need a GPU. Good.
Running in colab
Not running in colab
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="learning-as-a-hedge-against-uncertainty">
<h1>1.3.4.1: Learning as a Hedge Against Uncertainty<a class="headerlink" href="#learning-as-a-hedge-against-uncertainty" title="Permalink to this headline">¶</a></h1>
<p>One argument for learning from an evolutionary perspective is that some aspects of the environment change rapidly and in inherently unpredictable ways. When the ‘correct’ phenotype or behaviour depends on these unpredictable aspects, determination of the mature phenotype through development and, analogously, behavior through learning should be more contingent on environmental context and less determined genetically. The following model and simulations will help to illustrate how within-lifetime adaptation, e.g. learning, can be selected over genetic determination when the environment changes rapidly and unpredictably enough and how full genetic determination is favoured in environments that change only very slowly.</p>
<p>We further abstract our strike-no-strike game from sequence 1.3.2, interpreting the genetic loci as specifying the presence or absence of various neural connections, both inhibitory and excitatory. These connections could be thought of as existing between the receptors on the antennae and the muscles responsible for the striking behavior of our lurk-and-strike predator (the interpretation is inspired by Nolan and Hinton, 1987, from which much of this sequence is drawn), but they could also be thought of as specifying the layout of any neural circuit responsible for some set of environmentally contingent behaviours. We assume that there is some ideal ‘target’ neural configuration, specified by a target bitstring <span class="math notranslate nohighlight">\(y\)</span> of length <span class="math notranslate nohighlight">\(n\)</span> and so in the simplest case our fitness score function, <span class="math notranslate nohighlight">\(z\)</span>, for a genome <span class="math notranslate nohighlight">\(g\)</span> (also of length <span class="math notranslate nohighlight">\(n\)</span>) is defined as</p>
<div class="math notranslate nohighlight">
\[ \begin{align}\begin{aligned}z(g) := \sum_{i=1}^n I(y_i = g_i)$$.\\Here $I(y_i = g_i)$ is an indicator function which takes the value one when $y_i = g_i$ (indicating a match), and zero when the condition is not met. Roughly, the interpretation here is that as the actual neural configuration approaches the ideal neural configuration the particular behaviour the neural circuit governs becomes better suited to the environment with the resulting impact on fitness score.\\In this highly simplified model 'learning' is implemented as follows (still drawing from Hinton and Nolan 1987). Instead of our genome consisting exclusively of loci specifying the presence or absence of a given connection in the neural circuit, we consider two paired bitstrings $g$ and $f$. While the $g$ bits retain their original interpretation as specifying connections, the bits in $f$ (for 'flexibility') indicate that the corresponding connection is subject to within-lifetime determination through 'learning'. So the $g$ values at a given locus $g_i$ are only relevant to the fitness of the organism if $f_i=0$ and the connection is genetically determined, otherwise if $f_i=1$ the value of $g_i$ is mostly irrelevant since the connection will be set by learning. (Note that this is a kind of epistasis where the impact of one gene is heavily mediated or modified by another.) Our simplified learning mechanism tries a random combination of settings on each learning trial. If the combination of the flexible and genetically specified parameters ever produces a payoff that is either better or worse than any previously experienced payoffs, connections are set and fixed at a value that produced the higher of the two experienced payoffs and this configuration is used for the remainder of the learning trials. If no change in payoffs is experienced trial and error learning continues. This implicitly assumes that the organism can "recognize" when it has achieved a relatively good or bad setting, and recall a previous 'good' setting. This recognition ability is required for effective learning. It's important to note that this model represents a highly simplified abstraction of neural adaptation, and that typical real-world neural learning is far more complex and nuanced. This model, while simplified, captures essential dynamics that help illustrate the interplay between evolution and learning, allowing us to focus on the high-level interaction between learning and evolution without the intricacies of more complex and realistic neural adaptation.\\The total fitness score $F(g)$ for each individual will be the cumulative total payoff over all the learning trials and is given by\\$$F(g,f) := \sum_{t=1}^T z(g'(t))\end{aligned}\end{align} \]</div>
<p>where <span class="math notranslate nohighlight">\(g'(t)\)</span> is the test parameters used on each learning trial <span class="math notranslate nohighlight">\(t\)</span> of <span class="math notranslate nohighlight">\(T\)</span> total learning trials and is determined as</p>
<div class="math notranslate nohighlight">
\[\begin{split} g'_i(t) = \begin{cases}
g_i &amp; \text{if } f_i = 0 \\
\text{Bernoulli}(0.5) &amp; \text{if } f_i = 1 \text{ and } t \leq t^* \\
g^*_i &amp; \text{if } f_i = 1 \text{ and } t &gt; t^*
\end{cases}
\end{split}\]</div>
<p>Here <span class="math notranslate nohighlight">\(t^*\)</span> is the first learning trial where there is variance in the payoff and <span class="math notranslate nohighlight">\(g^*\)</span> are the settings associated with a high experienced payoff. So, in this model, <span class="math notranslate nohighlight">\(g^*\)</span> represents an optimal connection configuration discovered during the learning trials up and including trial <span class="math notranslate nohighlight">\(t^*\)</span> which is then consistently applied in subsequent trials now learning has stopped. Note that <span class="math notranslate nohighlight">\(F(g,f)\)</span> is inherently stochastic. Also, recall that different forms of selection, e.g. proportional (additive), truncation, proportional (multiplicative a.k.a. softmax), will translate the individuals’ fitness scores into fitness in the sense of replication rates.</p>
<p>In the coding exercise below we will implement the target based scoring function, and the learning rule, specified by the two aspects of the genome <span class="math notranslate nohighlight">\(g\)</span> and <span class="math notranslate nohighlight">\(f\)</span>. To explore this idea of learning being adaptive in variable environments we will also implement environmental change, operationalized by some proportion of the ‘target’ being subject to change over time. Specifically some fixed proportion of target bitstring <span class="math notranslate nohighlight">\(y\)</span> will have a fixed per-bit probability of flipping. This approach gives us a controlled way to examine how different rates of environmental change impact the balance between genetic determination and learning. In previous sequences we saw how this kind of dynamic environment is sometimes emergent amongst co-evolving populations, but here for simplicity we assume change just happens at a given rate (exogenously) without explicitly modeling the underlying evolutionary (or celestial or geological) drivers of this change.</p>
<div class="section" id="coding-exercise-implement-minimal-trial-and-error-learning">
<h2>Coding Exercise: Implement Minimal Trial and Error Learning<a class="headerlink" href="#coding-exercise-implement-minimal-trial-and-error-learning" title="Permalink to this headline">¶</a></h2>
<p>On each learning round <span class="math notranslate nohighlight">\(t\)</span> each individual who has yet to experience variable payoffs tries a new neural parameter configuration distributed as <span class="math notranslate nohighlight">\(g'(t)\)</span> above, and receives a resulting score <span class="math notranslate nohighlight">\(z(g'(t))\)</span>. Additionally, each round some proportion of the target <span class="math notranslate nohighlight">\(y\)</span> is subject to change with some fixed probability. Let’s adapt our evolutionary simulation function from the previous sequences to implement this simple form of learning and environmental change:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">################################################################################</span>
<span class="c1"># TODO for students: Complete the lines with ... that are wrapped in lines</span>
<span class="c1"># of # ~~~~ to implement this learning rule within our evolutionary simulation.</span>
<span class="c1"># Note that sometimes ellipsis '...' are used when indexing numpy arrays,</span>
<span class="c1"># these ellipsis are not meant to be changed or edited as part of the exercise</span>
<span class="c1"># Also, note that the simulation function is long and complex. Often when</span>
<span class="c1"># programming the primary challenge is not writing the code, but finding where</span>
<span class="c1"># the code needs a very small change. This exercise aims at developing this</span>
<span class="c1"># ability.</span>
<span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">"Exercise: implement simple learning rule"</span><span class="p">)</span>
<span class="c1">################################################################################</span>


<span class="k">def</span> <span class="nf">evo_learning_simulation</span><span class="p">(</span>
  <span class="n">pop_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">n_gen</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">num_blocks</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">block_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
  <span class="n">has_changing_environment</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">change_env_prop</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">env_change_rate</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span>
  <span class="n">has_learning</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_learning_trials</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
  <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">0.025</span><span class="p">,</span> <span class="n">has_mutation</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
  <span class="n">has_recombination</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">has_assortative_pairing</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">recombination_type</span><span class="o">=</span><span class="s1">'crossover'</span><span class="p">,</span>
  <span class="n">selection_type</span><span class="o">=</span><span class="s1">'softmax'</span><span class="p">,</span> <span class="c1">#softmax, deterministic truncation, proportional truncation</span>
  <span class="n">softmax_temp</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">truncation_threshold</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
  <span class="n">proportion_shift</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">proportion_scale</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
  <span class="n">seed</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">init_pop</span><span class="o">=</span><span class="s1">'simple'</span><span class="p">,</span>
  <span class="n">compute_IQR</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">  </span><span class="sd">"""</span>
<span class="sd">  Simulates the evolutionary process in a population of genotypes. This function</span>
<span class="sd">  models the evolution of genotypes through various mechanisms like learning,</span>
<span class="sd">  selection, mutation, recombination, and environmental changes. It is intended</span>
<span class="sd">  to explore how populations adapt to dynamic environments.</span>

<span class="sd">  Args:</span>
<span class="sd">    pop_size (int): Size of the population.</span>
<span class="sd">    n_gen (int): Number of generations to simulate.</span>
<span class="sd">    num_blocks (int): Number of receptors/genetic traits in each individual.</span>
<span class="sd">    has_changing_environment (bool): If True, the environment changes over generations.</span>
<span class="sd">    change_env_prop (float): Proportion of the environment that is subject to change.</span>
<span class="sd">    env_change_rate (float): Rate at which the environment changes.</span>
<span class="sd">    has_learning (bool): If True, individuals can learn during their lifetime.</span>
<span class="sd">    n_learning_trials (int): Number of learning trials per individual.</span>
<span class="sd">    mutation_rate (float): Rate of mutation in the population.</span>
<span class="sd">    has_mutation (bool): If True, mutations occur in the population.</span>
<span class="sd">    has_recombination (bool): If True, recombination occurs during reproduction.</span>
<span class="sd">    has_assortative_pairing (bool): If True, assortative mating is used in recombination.</span>
<span class="sd">    selection_type (str): Type of selection mechanism used.</span>
<span class="sd">    softmax_temp (float): Temperature parameter for softmax selection.</span>
<span class="sd">    truncation_threshold (float): Threshold for truncation selection.</span>
<span class="sd">    proportion_shift (float): Shift parameter for proportional selection.</span>
<span class="sd">    proportion_scale (float): Scale parameter for proportional selection.</span>
<span class="sd">    seed (int): Seed for random number generator.</span>
<span class="sd">    init_pop (str): Type of initial population ('simple' or 'adapted').</span>
<span class="sd">    compute_IQR (bool): If True, compute the interquartile range of the population scores</span>

<span class="sd">  Returns:</span>
<span class="sd">    dict: A dictionary containing historical data of various statistics over generations.</span>
<span class="sd">  """</span>
  <span class="c1"># Set seed</span>
  <span class="n">rng</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>

  <span class="k">if</span> <span class="ow">not</span> <span class="n">has_learning</span><span class="p">:</span>
    <span class="c1"># with only one trial there is no learning</span>
    <span class="n">n_learning_trials</span> <span class="o">=</span> <span class="mi">1</span>

  <span class="c1"># initialize the target connection pattern and how it changes each generation</span>
  <span class="n">base_genome_shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">num_blocks</span><span class="p">,</span> <span class="n">block_size</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">has_changing_environment</span><span class="p">:</span>
    <span class="n">env_target</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">base_genome_shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
    <span class="n">num_elements</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">base_genome_shape</span><span class="p">)</span>
    <span class="n">env_change_mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">base_genome_shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
    <span class="n">num_changes</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">change_env_prop</span> <span class="o">*</span> <span class="n">num_elements</span><span class="p">)</span>
    <span class="n">change_indices_1d</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">num_elements</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">num_changes</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">change_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unravel_index</span><span class="p">(</span><span class="n">change_indices_1d</span><span class="p">,</span> <span class="n">base_genome_shape</span><span class="p">)</span>
    <span class="n">env_change_mask</span><span class="p">[</span><span class="n">change_indices</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="n">env_target</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">base_genome_shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
    <span class="n">env_change_mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">base_genome_shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>

  <span class="n">genome_shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">pop_size</span><span class="p">,)</span> <span class="o">+</span> <span class="n">base_genome_shape</span> <span class="o">+</span> <span class="p">(</span><span class="mi">2</span><span class="p">,)</span>
  <span class="c1"># the two in the last dim is for preset and flexible</span>

  <span class="c1"># initialize the population genotypes</span>
  <span class="k">if</span> <span class="n">init_pop</span> <span class="o">==</span> <span class="s1">'simple'</span><span class="p">:</span>
    <span class="n">g_presets</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">genome_shape</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
    <span class="n">g_flexible</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">genome_shape</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
  <span class="k">elif</span> <span class="n">init_pop</span> <span class="o">==</span> <span class="s1">'adapted'</span><span class="p">:</span>
    <span class="c1"># For an adapted population, presets match the current environment</span>
    <span class="n">reps</span> <span class="o">=</span> <span class="p">[</span><span class="n">pop_size</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">env_target</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="n">g_presets</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">env_target</span><span class="p">,</span> <span class="n">reps</span><span class="p">)</span>  <span class="c1"># Replicate env_target for each individual</span>
    <span class="n">g_flexible</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">genome_shape</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
  <span class="n">population_genome</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">stack</span><span class="p">((</span><span class="n">g_presets</span><span class="p">,</span> <span class="n">g_flexible</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">genome_shape</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="c1"># pop_size x receptors x x block_size x 2</span>

  <span class="c1"># ensure even divisors/multiples for deterministic truncation</span>
  <span class="k">if</span> <span class="n">selection_type</span> <span class="o">==</span> <span class="s1">'deterministic truncation'</span><span class="p">:</span>
    <span class="n">num_parents</span> <span class="o">=</span> <span class="n">pop_size</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">truncation_threshold</span><span class="p">)</span>
    <span class="n">offspring_per_parent</span> <span class="o">=</span> <span class="n">pop_size</span> <span class="o">/</span> <span class="n">num_parents</span>
    <span class="c1"># Check if the numbers are close to integers, and even number of parents</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">isclose</span><span class="p">(</span><span class="n">num_parents</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">num_parents</span><span class="p">))</span> <span class="ow">and</span>
        <span class="n">np</span><span class="o">.</span><span class="n">isclose</span><span class="p">(</span><span class="n">offspring_per_parent</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">offspring_per_parent</span><span class="p">))</span> <span class="ow">and</span>
        <span class="n">num_parents</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
      <span class="nb">print</span><span class="p">(</span><span class="n">truncation_threshold</span><span class="p">)</span>
      <span class="nb">print</span><span class="p">(</span><span class="n">pop_size</span><span class="p">)</span>
      <span class="nb">print</span><span class="p">(</span><span class="n">num_parents</span><span class="p">)</span>
      <span class="nb">print</span><span class="p">(</span><span class="n">offspring_per_parent</span><span class="p">)</span>
      <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"For deterministic truncation, num_parents must be an even integer and divisor of pop_size, and offspring_per_parent must be an integer"</span><span class="p">)</span>
    <span class="n">num_parents</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">num_parents</span><span class="p">)</span>
    <span class="n">offspring_per_parent</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">offspring_per_parent</span><span class="p">)</span>

  <span class="c1"># Track statistics genotypes and scores over generations.</span>
  <span class="c1"># Note, depending on simulation type not all of these will be</span>
  <span class="c1"># updated and returned</span>
  <span class="n">mean_good_bits_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">mean_score_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">mean_flexible_change_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">mean_flexible_no_change_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">var_score_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">skew_score_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">lower_score_quartile_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">upper_score_quartile_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>

  <span class="c1"># Helper function calculate score of params against the environmental target</span>
  <span class="k">def</span> <span class="nf">score_calc</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">target</span><span class="p">):</span>
    <span class="c1"># params has shape pop_size x num_blocks x block_size</span>
    <span class="n">scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">params</span> <span class="o">==</span> <span class="n">target</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">2</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">scores</span>

  <span class="c1"># Run the simulation</span>
  <span class="k">for</span> <span class="n">generation</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_gen</span><span class="p">):</span>
    <span class="c1"># Calculate scores for each genotype</span>
    <span class="n">g_presets</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="o">...</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
    <span class="n">g_flexible</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="o">...</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">is_fixed</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">pop_size</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">has_learning</span><span class="p">:</span>
      <span class="n">best_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">pop_size</span><span class="p">)</span>
      <span class="n">cumulative_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">pop_size</span><span class="p">)</span>
      <span class="n">best_params</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">g_presets</span><span class="p">)</span>  <span class="c1"># Initially, best parameters are the presets</span>
      <span class="n">composite_params</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">g_presets</span><span class="p">)</span>  <span class="c1"># Initially, composite parameters are the presets</span>
      <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_learning_trials</span><span class="p">):</span>
        <span class="c1"># Explore only for individuals still learning</span>
        <span class="n">still_learning</span> <span class="o">=</span> <span class="o">~</span><span class="n">is_fixed</span>
        <span class="n">expanded_shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">pop_size</span><span class="p">,)</span> <span class="o">+</span> <span class="p">(</span><span class="mi">1</span><span class="p">,)</span> <span class="o">*</span> <span class="p">(</span><span class="n">g_flexible</span><span class="o">.</span><span class="n">ndim</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">still_learning_expanded</span> <span class="o">=</span> <span class="n">still_learning</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">expanded_shape</span><span class="p">)</span>
        <span class="c1"># This line is part of the coding exercise ~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
        <span class="c1"># update flexible params where learning is still happening</span>
        <span class="n">to_update</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
        <span class="c1"># ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
        <span class="c1"># Generate new samples only for the part of composite_params to be updated</span>
        <span class="n">update_param_samples</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">to_update</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
        <span class="c1"># This line is part of the coding exercise ~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
        <span class="n">composite_params</span><span class="p">[</span><span class="n">to_update</span><span class="p">]</span> <span class="o">=</span> <span class="o">...</span>
        <span class="c1"># ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
        <span class="c1"># trial scores are computed for everyone</span>
        <span class="n">trial_scores</span> <span class="o">=</span> <span class="n">score_calc</span><span class="p">(</span><span class="n">composite_params</span><span class="p">,</span> <span class="n">env_target</span><span class="p">)</span>
        <span class="c1"># but only learners should improve of decline</span>
        <span class="n">improved</span> <span class="o">=</span> <span class="n">trial_scores</span> <span class="o">&gt;</span> <span class="n">best_scores</span>
        <span class="n">declined</span> <span class="o">=</span> <span class="n">trial_scores</span> <span class="o">&lt;</span> <span class="n">best_scores</span>
        <span class="n">best_scores</span><span class="p">[</span><span class="n">improved</span><span class="p">]</span> <span class="o">=</span> <span class="n">trial_scores</span><span class="p">[</span><span class="n">improved</span><span class="p">]</span>
        <span class="n">best_params</span><span class="p">[</span><span class="n">improved</span><span class="p">]</span> <span class="o">=</span> <span class="n">composite_params</span><span class="p">[</span><span class="n">improved</span><span class="p">]</span>  <span class="c1"># Update best parameters</span>
        <span class="c1"># These lines are part of the coding exercise ~~~~~~~~~~~~~~~~~~~~~~~~~</span>
        <span class="c1"># our learning rule is stop after any improvement</span>
        <span class="n">is_fixed</span><span class="p">[</span><span class="n">improved</span><span class="p">]</span> <span class="o">=</span> <span class="o">...</span>
        <span class="c1"># or if things get worse stop exploring and use the known best</span>
        <span class="n">is_fixed</span><span class="p">[</span><span class="n">declined</span><span class="p">]</span> <span class="o">=</span> <span class="o">...</span>
        <span class="c1"># ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
        <span class="n">composite_params</span><span class="p">[</span><span class="n">declined</span><span class="p">]</span> <span class="o">=</span> <span class="n">best_params</span><span class="p">[</span><span class="n">declined</span><span class="p">]</span>
        <span class="n">cumulative_scores</span> <span class="o">+=</span> <span class="n">trial_scores</span>
      <span class="n">scores</span> <span class="o">=</span> <span class="n">cumulative_scores</span> <span class="o">/</span> <span class="n">n_learning_trials</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">scores</span> <span class="o">=</span> <span class="n">score_calc</span><span class="p">(</span><span class="n">g_presets</span><span class="p">,</span> <span class="n">env_target</span><span class="p">)</span>

    <span class="c1"># Track statistics of genotype scores over generations.</span>
    <span class="n">mean_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
    <span class="n">var_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
      <span class="n">skew_</span> <span class="o">=</span> <span class="p">((</span><span class="n">pop_size</span> <span class="o">/</span> <span class="p">(</span><span class="n">pop_size</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">pop_size</span><span class="o">-</span><span class="mi">2</span><span class="p">))</span> <span class="o">*</span>
        <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">scores</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">scores</span><span class="p">))</span><span class="o">**</span><span class="mi">3</span><span class="p">)</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">scores</span><span class="p">)))</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">skew_</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="n">flexible_env_and_g</span> <span class="o">=</span> <span class="n">g_flexible</span><span class="p">[:,</span> <span class="n">env_change_mask</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">dim</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">dim</span> <span class="ow">in</span> <span class="n">flexible_env_and_g</span><span class="o">.</span><span class="n">shape</span><span class="p">):</span>
      <span class="n">mean_flexible_change_</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">mean_flexible_change_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">flexible_env_and_g</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>

    <span class="n">fixed_env_flex_g</span> <span class="o">=</span> <span class="n">g_flexible</span><span class="p">[:,</span> <span class="o">~</span><span class="n">env_change_mask</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">dim</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">dim</span> <span class="ow">in</span> <span class="n">fixed_env_flex_g</span><span class="o">.</span><span class="n">shape</span><span class="p">):</span>
      <span class="n">mean_flexible_no_change_</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">mean_flexible_no_change_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">fixed_env_flex_g</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>

    <span class="n">mean_good_bits_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">g_presets</span> <span class="o">==</span> <span class="n">env_target</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">mean_score_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean_</span>
    <span class="n">var_score_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">var_</span>
    <span class="n">skew_score_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">skew_</span>
    <span class="n">lower_score_quartile_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="mi">25</span><span class="p">)</span>
    <span class="n">upper_score_quartile_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="mi">75</span><span class="p">)</span>
    <span class="n">mean_flexible_change_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean_flexible_change_</span>
    <span class="n">mean_flexible_no_change_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean_flexible_no_change_</span>
    <span class="n">mean_good_bits_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean_good_bits_</span>

    <span class="c1"># Selective Reproduction of different types</span>
    <span class="k">if</span> <span class="n">has_recombination</span> <span class="ow">and</span> <span class="n">has_assortative_pairing</span><span class="p">:</span>
      <span class="c1"># Shuffle and then sort the population by scores</span>
      <span class="n">shuffled_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">pop_size</span><span class="p">)</span>
      <span class="n">rng</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">shuffled_indices</span><span class="p">)</span>
      <span class="n">population_genome</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="n">shuffled_indices</span><span class="p">]</span>
      <span class="n">scores</span> <span class="o">=</span> <span class="n">scores</span><span class="p">[</span><span class="n">shuffled_indices</span><span class="p">]</span>
      <span class="c1"># Sort the population by scores</span>
      <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">scores</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>  <span class="c1"># Higher scores are better</span>
      <span class="n">population_genome</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
      <span class="n">scores</span> <span class="o">=</span> <span class="n">scores</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">selection_type</span> <span class="o">==</span> <span class="s1">'proportional'</span><span class="p">:</span>
      <span class="n">pos_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">scores</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">scores</span><span class="p">)</span>
      <span class="n">scale_shift_scores</span> <span class="o">=</span> <span class="p">(</span><span class="n">pos_scores</span> <span class="o">*</span> <span class="n">proportion_scale</span><span class="p">)</span> <span class="o">+</span> <span class="n">proportion_shift</span>
      <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">prob_scores</span> <span class="o">=</span> <span class="n">scale_shift_scores</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">scale_shift_scores</span><span class="p">)</span>
      <span class="k">else</span><span class="p">:</span>
        <span class="n">prob_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones_like</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">selection_type</span> <span class="o">==</span> <span class="s1">'softmax'</span><span class="p">:</span>
      <span class="n">stabilized_scores</span> <span class="o">=</span> <span class="n">scores</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
      <span class="n">exp_scaled_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">stabilized_scores</span> <span class="o">/</span> <span class="n">softmax_temp</span><span class="p">)</span>
      <span class="n">prob_scores</span> <span class="o">=</span> <span class="n">exp_scaled_scores</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">exp_scaled_scores</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">selection_type</span> <span class="o">==</span> <span class="s1">'proportional truncation'</span><span class="p">:</span>
      <span class="n">pos_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">scores</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">scores</span><span class="p">)</span>
      <span class="n">trunc_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
      <span class="n">selected</span> <span class="o">=</span> <span class="n">scores</span> <span class="o">&gt;=</span> <span class="n">np</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">truncation_threshold</span><span class="p">)</span>
      <span class="n">trunc_scores</span><span class="p">[</span><span class="n">selected</span><span class="p">]</span> <span class="o">=</span> <span class="n">pos_scores</span><span class="p">[</span><span class="n">selected</span><span class="p">]</span>
      <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">trunc_scores</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">prob_scores</span> <span class="o">=</span> <span class="n">trunc_scores</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">trunc_scores</span><span class="p">)</span>
      <span class="k">else</span><span class="p">:</span>
        <span class="n">prob_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones_like</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">selection_type</span> <span class="o">==</span> <span class="s1">'deterministic truncation'</span><span class="p">:</span>
      <span class="n">threshold_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">truncation_threshold</span><span class="p">)</span>
      <span class="c1"># Indices of individuals who meet or exceed the threshold score</span>
      <span class="n">eligible_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">scores</span> <span class="o">&gt;=</span> <span class="n">threshold_score</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
      <span class="c1"># If there are more eligible individuals than needed, take only the top</span>
      <span class="c1"># num_parents individuals</span>
      <span class="n">selected_indices</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">eligible_indices</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">scores</span><span class="p">[</span><span class="n">x</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)[:</span><span class="n">num_parents</span><span class="p">]</span>
      <span class="c1"># split these into parent groups</span>
      <span class="n">parent_set_1</span> <span class="o">=</span> <span class="n">selected_indices</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
      <span class="n">parent_set_2</span> <span class="o">=</span> <span class="n">selected_indices</span><span class="p">[</span><span class="mi">1</span><span class="p">::</span><span class="mi">2</span><span class="p">]</span>
      <span class="c1"># Replicate each parent in their respective sets</span>
      <span class="n">parent_set_1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="n">parent_set_1</span><span class="p">,</span> <span class="n">offspring_per_parent</span><span class="p">)</span>
      <span class="n">parent_set_2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="n">parent_set_2</span><span class="p">,</span> <span class="n">offspring_per_parent</span><span class="p">)</span>
      <span class="c1"># Shuffle the parent sets based on mating strategy</span>
      <span class="k">if</span> <span class="ow">not</span> <span class="n">has_assortative_pairing</span><span class="p">:</span>
        <span class="n">rng</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">parent_set_2</span><span class="p">)</span>
      <span class="c1"># Combine the two parent sets to form genome pairs</span>
      <span class="n">parent_pairs_shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,)</span> <span class="o">+</span> <span class="n">env_target</span><span class="o">.</span><span class="n">shape</span> <span class="o">+</span> <span class="p">(</span><span class="mi">2</span><span class="p">,)</span>
      <span class="n">parent_pairs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">parent_pairs_shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
      <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="n">parent_set_1</span><span class="p">]</span>
      <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="n">parent_set_2</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"Invalid selection_type string, use one of proportional, softmax, proportional truncation, or deterministic truncation"</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">selection_type</span> <span class="o">!=</span> <span class="s1">'deterministic truncation'</span><span class="p">:</span>
      <span class="c1"># form parent pairs</span>
      <span class="n">parent_pairs_shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,)</span> <span class="o">+</span> <span class="n">env_target</span><span class="o">.</span><span class="n">shape</span> <span class="o">+</span> <span class="p">(</span><span class="mi">2</span><span class="p">,)</span>
      <span class="n">parent_pairs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">parent_pairs_shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>

      <span class="k">if</span> <span class="n">has_recombination</span><span class="p">:</span>
        <span class="n">selected_parent_indices_1</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">pop_size</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span>
                                               <span class="n">p</span><span class="o">=</span><span class="n">prob_scores</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">pair_directions</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span> <span class="o">-</span> <span class="mi">1</span>  <span class="c1"># Results in either -1 or 1</span>
        <span class="k">if</span> <span class="n">has_assortative_pairing</span><span class="p">:</span>
          <span class="c1"># Use pair direction for assortative mating</span>
          <span class="n">selected_parent_indices_2</span> <span class="o">=</span> <span class="n">selected_parent_indices_1</span> <span class="o">+</span> <span class="n">pair_directions</span>
        <span class="k">else</span><span class="p">:</span>
          <span class="c1"># For non-assortative mating, select a second set of parents</span>
          <span class="n">selected_parent_indices_2</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">pop_size</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span>
                                                 <span class="n">p</span><span class="o">=</span><span class="n">prob_scores</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
          <span class="c1"># Correct self-pairing</span>
          <span class="n">selected_parent_indices_2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span> <span class="c1">#ternary use of where</span>
            <span class="n">selected_parent_indices_1</span> <span class="o">==</span> <span class="n">selected_parent_indices_2</span><span class="p">,</span> <span class="c1"># if this condition</span>
            <span class="p">(</span><span class="n">selected_parent_indices_2</span> <span class="o">+</span> <span class="n">pair_directions</span><span class="p">)</span> <span class="o">%</span> <span class="n">pop_size</span><span class="p">,</span> <span class="c1"># do this</span>
            <span class="n">selected_parent_indices_2</span><span class="p">)</span> <span class="c1"># other wise do this</span>
        <span class="c1"># Adjust any out-of-bounds indices</span>
        <span class="n">selected_parent_indices_2</span><span class="p">[</span><span class="n">selected_parent_indices_2</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">selected_parent_indices_2</span><span class="p">[</span><span class="n">selected_parent_indices_2</span> <span class="o">&gt;=</span> <span class="n">pop_size</span><span class="p">]</span> <span class="o">=</span> <span class="n">pop_size</span> <span class="o">-</span> <span class="mi">1</span>
      <span class="k">else</span><span class="p">:</span> <span class="c1"># no recombination</span>
        <span class="n">selected_parent_indices_1</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">pop_size</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span>
                                               <span class="n">p</span><span class="o">=</span><span class="n">prob_scores</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">selected_parent_indices_2</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">pop_size</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span>
                                               <span class="n">p</span><span class="o">=</span><span class="n">prob_scores</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
      <span class="c1"># use the selected indices to form the parent genome pairs</span>
      <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="n">selected_parent_indices_1</span><span class="p">]</span>
      <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="n">selected_parent_indices_2</span><span class="p">]</span>

    <span class="c1"># recombination</span>
    <span class="k">if</span> <span class="n">has_recombination</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">recombination_type</span> <span class="o">==</span> <span class="s1">'random'</span><span class="p">:</span>
        <span class="n">mask_shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,)</span> <span class="o">+</span> <span class="n">env_target</span><span class="o">.</span><span class="n">shape</span> <span class="o">+</span> <span class="p">(</span><span class="mi">2</span><span class="p">,)</span>
        <span class="n">mask1</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">mask_shape</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
        <span class="n">mask2</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">mask_shape</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
        <span class="n">children</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty_like</span><span class="p">(</span><span class="n">parent_pairs</span><span class="p">)</span>
        <span class="n">children</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask1</span><span class="p">,</span> <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span> <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">...</span><span class="p">])</span>
        <span class="n">children</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask2</span><span class="p">,</span> <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span> <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">...</span><span class="p">])</span>
      <span class="k">elif</span> <span class="n">recombination_type</span> <span class="o">==</span> <span class="s1">'crossover'</span><span class="p">:</span>
        <span class="c1"># Flatten the genomes for crossover operations</span>
        <span class="c1"># Shape before flattening: (pop_size // 2, 2, num_blocks, block_size, 2)</span>
        <span class="n">flat_parent_pairs</span> <span class="o">=</span> <span class="n">parent_pairs</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
        <span class="c1"># Shape after flattening: (pop_size // 2, 2, num_blocks * block_size * 2)</span>
        <span class="c1"># Generate two sets of unique crossover points for each parent pair</span>
        <span class="n">crossover_points_1</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">flat_parent_pairs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,))</span>
        <span class="n">crossover_points_2</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">flat_parent_pairs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,))</span>
        <span class="c1"># Create a range array that matches the last dimension of flat_parent_pairs</span>
        <span class="c1"># to help make the mask for crossover operations</span>
        <span class="n">range_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">flat_parent_pairs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1"># Use broadcasting to create masks: True if index is less than the crossover point</span>
        <span class="n">mask1</span> <span class="o">=</span> <span class="n">range_array</span> <span class="o">&lt;</span> <span class="n">crossover_points_1</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">]</span>
        <span class="n">mask2</span> <span class="o">=</span> <span class="n">range_array</span> <span class="o">&lt;</span> <span class="n">crossover_points_2</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">]</span>
        <span class="n">mask2</span> <span class="o">=</span> <span class="o">~</span><span class="n">mask2</span>
        <span class="n">children_flat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty_like</span><span class="p">(</span><span class="n">flat_parent_pairs</span><span class="p">)</span>
        <span class="n">children_flat</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask1</span><span class="p">,</span> <span class="n">flat_parent_pairs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">flat_parent_pairs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:])</span>
        <span class="n">children_flat</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask2</span><span class="p">,</span> <span class="n">flat_parent_pairs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">flat_parent_pairs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:])</span>
        <span class="n">children</span> <span class="o">=</span> <span class="n">children_flat</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">parent_pairs</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
      <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"Invalid recombination_type string, use one of random or crossover"</span><span class="p">)</span>
      <span class="n">recombined_population_genome</span> <span class="o">=</span> <span class="n">children</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">genome_shape</span><span class="p">)</span>
      <span class="n">population_genome</span> <span class="o">=</span> <span class="n">recombined_population_genome</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="c1"># the population is just the selected parents</span>
      <span class="n">population_genome</span> <span class="o">=</span> <span class="n">parent_pairs</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">genome_shape</span><span class="p">)</span>

    <span class="c1"># mutation</span>
    <span class="k">if</span> <span class="n">has_mutation</span><span class="p">:</span>
      <span class="n">mutation_mask</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">population_genome</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">mutation_rate</span>
      <span class="n">population_genome</span> <span class="o">^=</span> <span class="n">mutation_mask</span>

    <span class="c1"># dynamic environment</span>
    <span class="k">if</span> <span class="n">has_changing_environment</span><span class="p">:</span>
      <span class="c1"># see which aspect of the environment change and update them</span>
      <span class="n">did_change</span> <span class="o">=</span> <span class="n">env_change_mask</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rng</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="n">env_change_mask</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">env_change_rate</span><span class="p">)</span>
      <span class="c1"># flip the bits where change happened</span>
      <span class="n">env_target</span><span class="p">[</span><span class="n">did_change</span><span class="p">]</span> <span class="o">=</span> <span class="o">~</span><span class="n">env_target</span><span class="p">[</span><span class="n">did_change</span><span class="p">]</span>

    <span class="c1"># repeat the loop for n_gen iterations</span>

  <span class="n">results</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">'mean_score_hist'</span><span class="p">:</span> <span class="n">mean_score_hist</span><span class="p">,</span>
    <span class="s1">'var_score_hist'</span><span class="p">:</span> <span class="n">var_score_hist</span><span class="p">,</span>
    <span class="s1">'skew_score_hist'</span><span class="p">:</span> <span class="n">skew_score_hist</span><span class="p">,</span>
    <span class="s1">'mean_good_bits_hist'</span><span class="p">:</span> <span class="n">mean_good_bits_hist</span>
  <span class="p">}</span>
  <span class="k">if</span> <span class="n">compute_IQR</span><span class="p">:</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">'lower_score_quartile_hist'</span><span class="p">]</span> <span class="o">=</span> <span class="n">lower_score_quartile_hist</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">'upper_score_quartile_hist'</span><span class="p">]</span> <span class="o">=</span> <span class="n">upper_score_quartile_hist</span><span class="p">,</span>
  <span class="k">if</span> <span class="n">has_changing_environment</span><span class="p">:</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">'mean_flexible_change_hist'</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean_flexible_change_hist</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">'mean_flexible_no_change_hist'</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean_flexible_no_change_hist</span>
  <span class="k">return</span> <span class="n">results</span>

<span class="c1"># Define a range of environmental change rates</span>
<span class="n">env_change_rates</span> <span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">]</span>

<span class="c1"># Initialize dictionaries to store the full histories</span>
<span class="n">score_histories</span> <span class="o">=</span> <span class="p">{</span><span class="n">rate</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">rate</span> <span class="ow">in</span> <span class="n">env_change_rates</span><span class="p">}</span>
<span class="n">flexible_change_histories</span> <span class="o">=</span> <span class="p">{</span><span class="n">rate</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">rate</span> <span class="ow">in</span> <span class="n">env_change_rates</span><span class="p">}</span>
<span class="n">flexible_no_change_histories</span> <span class="o">=</span> <span class="p">{</span><span class="n">rate</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">rate</span> <span class="ow">in</span> <span class="n">env_change_rates</span><span class="p">}</span>

<span class="c1"># Run simulations for each change rate</span>
<span class="k">for</span> <span class="n">rate</span> <span class="ow">in</span> <span class="n">env_change_rates</span><span class="p">:</span>
  <span class="n">simulation_results</span> <span class="o">=</span> <span class="n">evo_learning_simulation</span><span class="p">(</span>
      <span class="n">env_change_rate</span><span class="o">=</span><span class="n">rate</span><span class="p">,</span>
    <span class="p">)</span>
  <span class="c1"># Store the full histories</span>
  <span class="n">score_histories</span><span class="p">[</span><span class="n">rate</span><span class="p">]</span> <span class="o">=</span> <span class="n">simulation_results</span><span class="p">[</span><span class="s1">'mean_score_hist'</span><span class="p">]</span>
  <span class="n">flexible_change_histories</span><span class="p">[</span><span class="n">rate</span><span class="p">]</span> <span class="o">=</span> <span class="n">simulation_results</span><span class="p">[</span><span class="s1">'mean_flexible_change_hist'</span><span class="p">]</span>
  <span class="n">flexible_no_change_histories</span><span class="p">[</span><span class="n">rate</span><span class="p">]</span> <span class="o">=</span> <span class="n">simulation_results</span><span class="p">[</span><span class="s1">'mean_flexible_no_change_hist'</span><span class="p">]</span>

<span class="c1"># Plotting using the object-oriented approach</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">))</span>

<span class="c1"># Plot Mean Score Histories</span>
<span class="k">for</span> <span class="n">rate</span><span class="p">,</span> <span class="n">history</span> <span class="ow">in</span> <span class="n">score_histories</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">history</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">'Change Rate </span><span class="si">{</span><span class="n">rate</span><span class="si">:</span><span class="s1">.2f</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">'Generation'</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">'Mean Score'</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">'Mean Score Over Generations'</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="c1"># Plot Mean Flexibility for Change Eligible Histories</span>
<span class="k">for</span> <span class="n">rate</span><span class="p">,</span> <span class="n">history</span> <span class="ow">in</span> <span class="n">flexible_change_histories</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">history</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">'Change Rate </span><span class="si">{</span><span class="n">rate</span><span class="si">:</span><span class="s1">.2f</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">'Generation'</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">'Mean Flexibility (Change Eligible)'</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">'Mean Flexibility (Change Eligible) Over Generations'</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="c1"># Plot Mean Flexibility for No Change Eligible Histories</span>
<span class="k">for</span> <span class="n">rate</span><span class="p">,</span> <span class="n">history</span> <span class="ow">in</span> <span class="n">flexible_no_change_histories</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">history</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">'Change Rate </span><span class="si">{</span><span class="n">rate</span><span class="si">:</span><span class="s1">.2f</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">'Generation'</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">'Mean Flexibility (No Change Eligible)'</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">'Mean Flexibility (No Change Eligible) Over Generations'</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="n">remove_ip_clutter</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>

</div>
<p><a class="reference external" href="https://github.com/dcownden/PerennialProblemsOfLifeWithABrain/tree/main//sequences/P1C3_RealEvolution/solutions/P1C3_Sequence4_Solution_9493df6d.py"><em>Click for solution</em></a></p>
<p><em>Example output:</em></p>
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/dcownden/PerennialProblemsOfLifeWithABrain/main/sequences/P1C3_RealEvolution/static/P1C3_Sequence4_Solution_9493df6d_0.png"><img align="center" alt="Solution hint" class="align-center" src="https://raw.githubusercontent.com/dcownden/PerennialProblemsOfLifeWithABrain/main/sequences/P1C3_RealEvolution/static/P1C3_Sequence4_Solution_9493df6d_0.png" style="width: 1000.0px; height: 1500.0px;"/></a>
<p>The simulation results above help illustrate how the adaptive balance between learning (plasticity) and genetic determination shifts in response to environmental variability.</p>
<p>In the first plot “Mean Score Over Generations”, we see a clear trend: stable environments lead to higher and more rapidly stabilizing mean scores (blue line, change rate = 0.00, orange line, change rate = 0.01). As the environment’s change rate increases (red and green lines), mean scores become more erratic and generally lower. This aligns with the expectation that optimization through evolution will always lag behind in dynamic environments where the ‘ideal phenotype’ can change from generation to generation.</p>
<p>The second plot, “Mean Flexibility (Change Eligible) Over Generations”, tracks the evolution of the mean proportion of learning-determined loci, for those loci associated with fluctuating environmental targets (there are 10 such loci in these particular simulations). The plot shows a higher mean flexibility at loci where environmental targets change frequently (green, 0.1, and red, 0.5, lines) compared to loci with slow or no change (orange, 0.01, and blue, 0.0, lines respectively). This suggests that learning is advantageous and selected for when environmental unpredictability is high, as it gives organisms a way to rapidly adjust to new conditions within their lifetime and thus reduces dependence on genetic changes over generational time to cope with environmental dynamics (so called adaptive tracking). When the environmental volatility is very high (red, 0.5) there is basically no evolutionary tracking of the variable environment, and learning/plasticity dominates. In contrast, at an intermediate level of volatility (green, 0.1) we see that a combination of adaptive tracking and plasticity is stable within the population.</p>
<p>The “Mean Flexibility (No Change Eligible) Over Generations” plot serves as a sanity check, confirming that where environmental targets don’t change, learning is selected against. When there is no environmental variability full genetic determination is selected for.</p>
<p>The overarching message from these simulations is clear: genetic determination is beneficial in static environments where the optimal phenotype is constant. In such cases the ‘correct phenotype’ can in some sense be known fully by the genome and so should simply be developed as rapidly as possible. In such cases learning and other forms of phenotypic plasticity that might slow development have no adaptive value. Conversely, in volatile environments, learning and plasticity can be selected for. Note that in these simulations all of the many mechanisms required for effective plasticity have been abstracted away, e.g. the details of how flexible development/neural plasticity, evaluation of different outcomes, sensing the environment, etc. is achieved is left completely unanswered and so these simulations serve only to hint at the kinds of scenarios when such mechanism could be evolutionarily favoured, without giving any consideration to the costs associated with developing and maintaining such mechanisms.</p>
<p>Submit your feedback</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># @markdown Submit your feedback</span>
<span class="n">content_review</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">feedback_prefix</span><span class="si">}</span><span class="s2">_M1"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/vnd.jupyter.widget-view+json">
{"version_major": 2, "version_minor": 0, "model_id": "4bacb49152ae4176b4715917629a73c0"}
</script></div>
</div>
</div>
</div>
<div class="section" id="some-fitness-score-functions-are-hard-for-evolution-to-optimize">
<h1>1.3.4.2: Some Fitness Score Functions are Hard for Evolution to Optimize<a class="headerlink" href="#some-fitness-score-functions-are-hard-for-evolution-to-optimize" title="Permalink to this headline">¶</a></h1>
<p>In the previous section we saw how evolutionary processes are responsive to (but not anticipatory of) environmental change, and how as a result, in a changing environment the population phenotype will often lag behind the optimal phenotype for a given environment. We also saw how within lifetime learning could be an adaptive response to dealing with environmental change and uncertainty. In the previous simulations the connection between a genetic change and the fitness increment was quite direct given the structure of the fitness score function (the sum of ‘correct’ alleles) and so as a result selection was able to (relatively) rapidly steer the population towards good phenotypes for the given environment. This is not always the case.</p>
<p>Genes rarely operate in isolation. They are typically part of dynamic networks, engaging in a molecular dialogue to orchestrate complex patterns of gene expression. These networks shape the diverse phenotypes (traits) we observe, from physical characteristics to behavioural patterns. Pleiotropy, where a single gene influences multiple traits, and epistasis, where the effect of one gene is significantly shaped by others, are the norm rather than the exception. Consequently, most traits are polygenic, emerging from the collective influence of multiple genes. Note that in our models we rarely consider these complications, and for the illustrative goals of most of our models we do not need to. However, if and when the goal of modeling becomes quantitative prediction actual empirical gene frequencies, these details, in particular when they can and cannot be ignored in modeling, do become critical.</p>
<p>For a moment now though we <em>are</em> going to consider how a very simplified form of epistasis can result in a fitness score function that is much more difficult for an evolutionary process to optimize.</p>
<p>We will continue to use the ‘strike-no-strike’ decision problem but with some added complexity. We still consider an organism with <span class="math notranslate nohighlight">\(n\)</span> chemosensory receptors, each corresponding to <span class="math notranslate nohighlight">\(n\)</span> possible food/non-food chemical markers in the environment. However, now the functionality of these receptors is determined by <span class="math notranslate nohighlight">\(nk\)</span> distinct genetic switches, represented as a bit-string <span class="math notranslate nohighlight">\(g\)</span> of length <span class="math notranslate nohighlight">\(nk\)</span> with every block of <span class="math notranslate nohighlight">\(k\)</span> contiguous bits determining the efficacy of a receptor in the strike-no-strike game. Instead of thinking of each genetic bit as corresponding to the expression or absence of a receptor for a particular marker, we are instead now thinking of each genetic bit as determining some connections in a neural circuit between receptors and muscles and only a very particular set of connections allows for the effective use of the receptor. In our learning simulations above we had a kind of epistasis where the the impact of the <span class="math notranslate nohighlight">\(g\)</span> (preset) type loci was dependent on the the <span class="math notranslate nohighlight">\(f\)</span> (flexibility) type loci. Here we have an even stronger form of epistasis where the fitness impact of any one <span class="math notranslate nohighlight">\(g\)</span> type loci is fully dependent on the other genes that share its receptor block.</p>
<p>In  the previous model having the ‘correct’ allele, i.e. <span class="math notranslate nohighlight">\(g_i = y_i\)</span>, at any loci, conferred a small, independent, additive benefit to fitness score. Now having the correct allele at a given loci, only provides a fitness score benefit if the other <span class="math notranslate nohighlight">\(k-1\)</span> loci in that same ‘receptor block’ on the genome also happen to be correct, otherwise the independent effect of that loci is neutral.</p>
<p>Making the same rough interpretation as before when a given block of connections aligns with the target neural circuit configuration the particular behaviour the neural circuit governs becomes better suited to the environment with the resulting impact on fitness score. This gives us our single trial fitness score as</p>
<div class="math notranslate nohighlight">
\[ z(g):= \sum_{i=0}^{n-1} \prod_{j=1}^{k} I(y_{j+ik} = g_{j+ik}) \]</div>
<div class="section" id="coding-exercise-implement-the-new-fitness-score-function">
<h2>Coding Exercise: Implement the New Fitness Score Function.<a class="headerlink" href="#coding-exercise-implement-the-new-fitness-score-function" title="Permalink to this headline">¶</a></h2>
<p>Each individual in the population has a genotype <span class="math notranslate nohighlight">\(g\)</span>, and a resulting score per trial <span class="math notranslate nohighlight">\(\sum_{i=0}^{n-1} \prod_{j=1}^{k} I(y_{j+ik} = g_{j+ik})\)</span> i.e. the number of blocks of entirely ‘good bits’. We will leave learning and environmental change out of our simulation in this first instance. Here <span class="math notranslate nohighlight">\(g\)</span> is the genetic specification of the neural circuit <span class="math notranslate nohighlight">\(y\)</span> is the target and <span class="math notranslate nohighlight">\(I\)</span> is an indicator function. Let’s adapt our evolutionary simulation function from before to use this new fitness score function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">################################################################################</span>
<span class="c1"># TODO for students: Complete the score_calc function definition to implement</span>
<span class="c1"># this new per-trial score function. Hints, 1) think hard about the shapes of</span>
<span class="c1"># the composite_params and env_target arrays, 2) use numpy broadcasting to</span>
<span class="c1"># compare all composite_params to the target 3) use numpy sum and prod functions</span>
<span class="c1"># along specific axis, 4) play around with small test arrays in other (new)</span>
<span class="c1"># code cells before implementing your solution.</span>
<span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">"Exercise: implement complex score function"</span><span class="p">)</span>
<span class="c1">################################################################################</span>


<span class="k">def</span> <span class="nf">evo_learning_simulation</span><span class="p">(</span>
  <span class="n">pop_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">n_gen</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">num_blocks</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">block_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
  <span class="n">has_changing_environment</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">change_env_prop</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">env_change_rate</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span>
  <span class="n">has_learning</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_learning_trials</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
  <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">0.025</span><span class="p">,</span> <span class="n">has_mutation</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
  <span class="n">has_recombination</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">has_assortative_pairing</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">recombination_type</span><span class="o">=</span><span class="s1">'crossover'</span><span class="p">,</span>
  <span class="n">selection_type</span><span class="o">=</span><span class="s1">'softmax'</span><span class="p">,</span> <span class="c1">#softmax, deterministic truncation, proportional truncation</span>
  <span class="n">softmax_temp</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">truncation_threshold</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
  <span class="n">proportion_shift</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">proportion_scale</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
  <span class="n">seed</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">init_pop</span><span class="o">=</span><span class="s1">'simple'</span><span class="p">,</span>
  <span class="n">compute_IQR</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">  </span><span class="sd">"""</span>
<span class="sd">  Simulates the evolutionary process in a population of genotypes. This function</span>
<span class="sd">  models the evolution of genotypes through various mechanisms like learning,</span>
<span class="sd">  selection, mutation, recombination, and environmental changes. It is intended</span>
<span class="sd">  to explore how populations adapt to dynamic environments.</span>

<span class="sd">  Args:</span>
<span class="sd">    pop_size (int): Size of the population.</span>
<span class="sd">    n_gen (int): Number of generations to simulate.</span>
<span class="sd">    num_blocks (int): Number of receptors/genetic traits in each individual.</span>
<span class="sd">    has_changing_environment (bool): If True, the environment changes over generations.</span>
<span class="sd">    change_env_prop (float): Proportion of the environment that is subject to change.</span>
<span class="sd">    env_change_rate (float): Rate at which the environment changes.</span>
<span class="sd">    has_learning (bool): If True, individuals can learn during their lifetime.</span>
<span class="sd">    n_learning_trials (int): Number of learning trials per individual.</span>
<span class="sd">    mutation_rate (float): Rate of mutation in the population.</span>
<span class="sd">    has_mutation (bool): If True, mutations occur in the population.</span>
<span class="sd">    has_recombination (bool): If True, recombination occurs during reproduction.</span>
<span class="sd">    has_assortative_pairing (bool): If True, assortative mating is used in recombination.</span>
<span class="sd">    selection_type (str): Type of selection mechanism used.</span>
<span class="sd">    softmax_temp (float): Temperature parameter for softmax selection.</span>
<span class="sd">    truncation_threshold (float): Threshold for truncation selection.</span>
<span class="sd">    proportion_shift (float): Shift parameter for proportional selection.</span>
<span class="sd">    proportion_scale (float): Scale parameter for proportional selection.</span>
<span class="sd">    seed (int): Seed for random number generator.</span>
<span class="sd">    init_pop (str): Type of initial population ('simple' or 'adapted').</span>
<span class="sd">    compute_IQR (bool): If True, compute the interquartile range of the population scores</span>

<span class="sd">  Returns:</span>
<span class="sd">    dict: A dictionary containing historical data of various statistics over generations.</span>
<span class="sd">  """</span>
  <span class="c1"># Set seed</span>
  <span class="n">rng</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>

  <span class="k">if</span> <span class="ow">not</span> <span class="n">has_learning</span><span class="p">:</span>
    <span class="c1"># with only one trial there is no learning</span>
    <span class="n">n_learning_trials</span> <span class="o">=</span> <span class="mi">1</span>

  <span class="c1"># initialize the target connection pattern and how it changes each generation</span>
  <span class="n">base_genome_shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">num_blocks</span><span class="p">,</span> <span class="n">block_size</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">has_changing_environment</span><span class="p">:</span>
    <span class="n">env_target</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">base_genome_shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
    <span class="n">num_elements</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">base_genome_shape</span><span class="p">)</span>
    <span class="n">env_change_mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">base_genome_shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
    <span class="n">num_changes</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">change_env_prop</span> <span class="o">*</span> <span class="n">num_elements</span><span class="p">)</span>
    <span class="n">change_indices_1d</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">num_elements</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">num_changes</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">change_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unravel_index</span><span class="p">(</span><span class="n">change_indices_1d</span><span class="p">,</span> <span class="n">base_genome_shape</span><span class="p">)</span>
    <span class="n">env_change_mask</span><span class="p">[</span><span class="n">change_indices</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="n">env_target</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">base_genome_shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
    <span class="n">env_change_mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">base_genome_shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>

  <span class="n">genome_shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">pop_size</span><span class="p">,)</span> <span class="o">+</span> <span class="n">base_genome_shape</span> <span class="o">+</span> <span class="p">(</span><span class="mi">2</span><span class="p">,)</span>
  <span class="c1"># the two in the last dim is for preset and flexible</span>

  <span class="c1"># initialize the population genotypes</span>
  <span class="k">if</span> <span class="n">init_pop</span> <span class="o">==</span> <span class="s1">'simple'</span><span class="p">:</span>
    <span class="n">g_presets</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">genome_shape</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
    <span class="n">g_flexible</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">genome_shape</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
  <span class="k">elif</span> <span class="n">init_pop</span> <span class="o">==</span> <span class="s1">'adapted'</span><span class="p">:</span>
    <span class="c1"># For an adapted population, presets match the current environment</span>
    <span class="n">reps</span> <span class="o">=</span> <span class="p">[</span><span class="n">pop_size</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">env_target</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="n">g_presets</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">env_target</span><span class="p">,</span> <span class="n">reps</span><span class="p">)</span>  <span class="c1"># Replicate env_target for each individual</span>
    <span class="n">g_flexible</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">genome_shape</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
  <span class="n">population_genome</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">stack</span><span class="p">((</span><span class="n">g_presets</span><span class="p">,</span> <span class="n">g_flexible</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">genome_shape</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="c1"># pop_size x receptors x x block_size x 2</span>

  <span class="c1"># ensure even divisors/multiples for deterministic truncation</span>
  <span class="k">if</span> <span class="n">selection_type</span> <span class="o">==</span> <span class="s1">'deterministic truncation'</span><span class="p">:</span>
    <span class="n">num_parents</span> <span class="o">=</span> <span class="n">pop_size</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">truncation_threshold</span><span class="p">)</span>
    <span class="n">offspring_per_parent</span> <span class="o">=</span> <span class="n">pop_size</span> <span class="o">/</span> <span class="n">num_parents</span>
    <span class="c1"># Check if the numbers are close to integers, and even number of parents</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">isclose</span><span class="p">(</span><span class="n">num_parents</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">num_parents</span><span class="p">))</span> <span class="ow">and</span>
        <span class="n">np</span><span class="o">.</span><span class="n">isclose</span><span class="p">(</span><span class="n">offspring_per_parent</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">offspring_per_parent</span><span class="p">))</span> <span class="ow">and</span>
        <span class="n">num_parents</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
      <span class="nb">print</span><span class="p">(</span><span class="n">truncation_threshold</span><span class="p">)</span>
      <span class="nb">print</span><span class="p">(</span><span class="n">pop_size</span><span class="p">)</span>
      <span class="nb">print</span><span class="p">(</span><span class="n">num_parents</span><span class="p">)</span>
      <span class="nb">print</span><span class="p">(</span><span class="n">offspring_per_parent</span><span class="p">)</span>
      <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"For deterministic truncation, num_parents must be an even integer and divisor of pop_size, and offspring_per_parent must be an integer"</span><span class="p">)</span>
    <span class="n">num_parents</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">num_parents</span><span class="p">)</span>
    <span class="n">offspring_per_parent</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">offspring_per_parent</span><span class="p">)</span>

  <span class="c1"># Track statistics genotypes and scores over generations.</span>
  <span class="c1"># Note, depending on simulation type not all of these will be</span>
  <span class="c1"># updated and returned</span>
  <span class="n">mean_good_bits_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">mean_score_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">mean_flexible_change_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">mean_flexible_no_change_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">var_score_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">skew_score_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">lower_score_quartile_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>
  <span class="n">upper_score_quartile_hist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_gen</span><span class="p">)</span>

  <span class="c1"># Helper function calculate score of params against the environmental target</span>
  <span class="k">def</span> <span class="nf">score_calc</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">target</span><span class="p">):</span>
    <span class="c1"># params has shape pop_size x num_blocks x block_size</span>
    <span class="n">scores</span> <span class="o">=</span> <span class="o">...</span>
    <span class="k">return</span> <span class="n">scores</span>

  <span class="c1"># Run the simulation</span>
  <span class="k">for</span> <span class="n">generation</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_gen</span><span class="p">):</span>
    <span class="c1"># Calculate scores for each genotype</span>
    <span class="n">g_presets</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="o">...</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
    <span class="n">g_flexible</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="o">...</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">is_fixed</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">pop_size</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">has_learning</span><span class="p">:</span>
      <span class="n">best_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">pop_size</span><span class="p">)</span>
      <span class="n">cumulative_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">pop_size</span><span class="p">)</span>
      <span class="n">best_params</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">g_presets</span><span class="p">)</span>  <span class="c1"># Initially, best parameters are the presets</span>
      <span class="n">composite_params</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">g_presets</span><span class="p">)</span>  <span class="c1"># Initially, composite parameters are the presets</span>
      <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_learning_trials</span><span class="p">):</span>
        <span class="c1"># Explore only for individuals still learning</span>
        <span class="n">still_learning</span> <span class="o">=</span> <span class="o">~</span><span class="n">is_fixed</span>
        <span class="n">expanded_shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">pop_size</span><span class="p">,)</span> <span class="o">+</span> <span class="p">(</span><span class="mi">1</span><span class="p">,)</span> <span class="o">*</span> <span class="p">(</span><span class="n">g_flexible</span><span class="o">.</span><span class="n">ndim</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">still_learning_expanded</span> <span class="o">=</span> <span class="n">still_learning</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">expanded_shape</span><span class="p">)</span>
        <span class="n">to_update</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">still_learning_expanded</span><span class="p">,</span> <span class="n">g_flexible</span><span class="p">)</span>
        <span class="c1"># Generate new samples only for the part of composite_params to be updated</span>
        <span class="n">update_param_samples</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">to_update</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
        <span class="n">composite_params</span><span class="p">[</span><span class="n">to_update</span><span class="p">]</span> <span class="o">=</span> <span class="n">update_param_samples</span>
        <span class="c1"># trial scores are computed for everyone</span>
        <span class="n">trial_scores</span> <span class="o">=</span> <span class="n">score_calc</span><span class="p">(</span><span class="n">composite_params</span><span class="p">,</span> <span class="n">env_target</span><span class="p">)</span>
        <span class="c1"># but only learners should improve of decline</span>
        <span class="n">improved</span> <span class="o">=</span> <span class="n">trial_scores</span> <span class="o">&gt;</span> <span class="n">best_scores</span>
        <span class="n">declined</span> <span class="o">=</span> <span class="n">trial_scores</span> <span class="o">&lt;</span> <span class="n">best_scores</span>
        <span class="n">best_scores</span><span class="p">[</span><span class="n">improved</span><span class="p">]</span> <span class="o">=</span> <span class="n">trial_scores</span><span class="p">[</span><span class="n">improved</span><span class="p">]</span>
        <span class="n">best_params</span><span class="p">[</span><span class="n">improved</span><span class="p">]</span> <span class="o">=</span> <span class="n">composite_params</span><span class="p">[</span><span class="n">improved</span><span class="p">]</span>  <span class="c1"># Update best parameters</span>
        <span class="c1"># our learning rule is stop after any improvement</span>
        <span class="n">is_fixed</span><span class="p">[</span><span class="n">improved</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="c1"># or if things get worse stop exploring and use the known best</span>
        <span class="n">is_fixed</span><span class="p">[</span><span class="n">declined</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">composite_params</span><span class="p">[</span><span class="n">declined</span><span class="p">]</span> <span class="o">=</span> <span class="n">best_params</span><span class="p">[</span><span class="n">declined</span><span class="p">]</span>
        <span class="n">cumulative_scores</span> <span class="o">+=</span> <span class="n">trial_scores</span>
      <span class="n">scores</span> <span class="o">=</span> <span class="n">cumulative_scores</span> <span class="o">/</span> <span class="n">n_learning_trials</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">scores</span> <span class="o">=</span> <span class="n">score_calc</span><span class="p">(</span><span class="n">g_presets</span><span class="p">,</span> <span class="n">env_target</span><span class="p">)</span>

    <span class="c1"># Track statistics of genotype scores over generations.</span>
    <span class="n">mean_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
    <span class="n">var_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
      <span class="n">skew_</span> <span class="o">=</span> <span class="p">((</span><span class="n">pop_size</span> <span class="o">/</span> <span class="p">(</span><span class="n">pop_size</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">pop_size</span><span class="o">-</span><span class="mi">2</span><span class="p">))</span> <span class="o">*</span>
        <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">scores</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">scores</span><span class="p">))</span><span class="o">**</span><span class="mi">3</span><span class="p">)</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">scores</span><span class="p">)))</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">skew_</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="n">flexible_env_and_g</span> <span class="o">=</span> <span class="n">g_flexible</span><span class="p">[:,</span> <span class="n">env_change_mask</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">dim</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">dim</span> <span class="ow">in</span> <span class="n">flexible_env_and_g</span><span class="o">.</span><span class="n">shape</span><span class="p">):</span>
      <span class="n">mean_flexible_change_</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">mean_flexible_change_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">flexible_env_and_g</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>

    <span class="n">fixed_env_flex_g</span> <span class="o">=</span> <span class="n">g_flexible</span><span class="p">[:,</span> <span class="o">~</span><span class="n">env_change_mask</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">dim</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">dim</span> <span class="ow">in</span> <span class="n">fixed_env_flex_g</span><span class="o">.</span><span class="n">shape</span><span class="p">):</span>
      <span class="n">mean_flexible_no_change_</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">mean_flexible_no_change_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">fixed_env_flex_g</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>

    <span class="n">mean_good_bits_</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">g_presets</span> <span class="o">==</span> <span class="n">env_target</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">mean_score_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean_</span>
    <span class="n">var_score_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">var_</span>
    <span class="n">skew_score_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">skew_</span>
    <span class="n">lower_score_quartile_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="mi">25</span><span class="p">)</span>
    <span class="n">upper_score_quartile_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="mi">75</span><span class="p">)</span>
    <span class="n">mean_flexible_change_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean_flexible_change_</span>
    <span class="n">mean_flexible_no_change_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean_flexible_no_change_</span>
    <span class="n">mean_good_bits_hist</span><span class="p">[</span><span class="n">generation</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean_good_bits_</span>

    <span class="c1"># Selective Reproduction of different types</span>
    <span class="k">if</span> <span class="n">has_recombination</span> <span class="ow">and</span> <span class="n">has_assortative_pairing</span><span class="p">:</span>
      <span class="c1"># Shuffle and then sort the population by scores</span>
      <span class="n">shuffled_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">pop_size</span><span class="p">)</span>
      <span class="n">rng</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">shuffled_indices</span><span class="p">)</span>
      <span class="n">population_genome</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="n">shuffled_indices</span><span class="p">]</span>
      <span class="n">scores</span> <span class="o">=</span> <span class="n">scores</span><span class="p">[</span><span class="n">shuffled_indices</span><span class="p">]</span>
      <span class="c1"># Sort the population by scores</span>
      <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">scores</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>  <span class="c1"># Higher scores are better</span>
      <span class="n">population_genome</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
      <span class="n">scores</span> <span class="o">=</span> <span class="n">scores</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">selection_type</span> <span class="o">==</span> <span class="s1">'proportional'</span><span class="p">:</span>
      <span class="n">pos_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">scores</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">scores</span><span class="p">)</span>
      <span class="n">scale_shift_scores</span> <span class="o">=</span> <span class="p">(</span><span class="n">pos_scores</span> <span class="o">*</span> <span class="n">proportion_scale</span><span class="p">)</span> <span class="o">+</span> <span class="n">proportion_shift</span>
      <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">prob_scores</span> <span class="o">=</span> <span class="n">scale_shift_scores</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">scale_shift_scores</span><span class="p">)</span>
      <span class="k">else</span><span class="p">:</span>
        <span class="n">prob_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones_like</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">selection_type</span> <span class="o">==</span> <span class="s1">'softmax'</span><span class="p">:</span>
      <span class="n">stabilized_scores</span> <span class="o">=</span> <span class="n">scores</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
      <span class="n">exp_scaled_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">stabilized_scores</span> <span class="o">/</span> <span class="n">softmax_temp</span><span class="p">)</span>
      <span class="n">prob_scores</span> <span class="o">=</span> <span class="n">exp_scaled_scores</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">exp_scaled_scores</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">selection_type</span> <span class="o">==</span> <span class="s1">'proportional truncation'</span><span class="p">:</span>
      <span class="n">pos_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">scores</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">scores</span><span class="p">)</span>
      <span class="n">trunc_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
      <span class="n">selected</span> <span class="o">=</span> <span class="n">scores</span> <span class="o">&gt;=</span> <span class="n">np</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">truncation_threshold</span><span class="p">)</span>
      <span class="n">trunc_scores</span><span class="p">[</span><span class="n">selected</span><span class="p">]</span> <span class="o">=</span> <span class="n">pos_scores</span><span class="p">[</span><span class="n">selected</span><span class="p">]</span>
      <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">trunc_scores</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">prob_scores</span> <span class="o">=</span> <span class="n">trunc_scores</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">trunc_scores</span><span class="p">)</span>
      <span class="k">else</span><span class="p">:</span>
        <span class="n">prob_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones_like</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">selection_type</span> <span class="o">==</span> <span class="s1">'deterministic truncation'</span><span class="p">:</span>
      <span class="n">threshold_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">truncation_threshold</span><span class="p">)</span>
      <span class="c1"># Indices of individuals who meet or exceed the threshold score</span>
      <span class="n">eligible_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">scores</span> <span class="o">&gt;=</span> <span class="n">threshold_score</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
      <span class="c1"># If there are more eligible individuals than needed, take only the top</span>
      <span class="c1"># num_parents individuals</span>
      <span class="n">selected_indices</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">eligible_indices</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">scores</span><span class="p">[</span><span class="n">x</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)[:</span><span class="n">num_parents</span><span class="p">]</span>
      <span class="c1"># split these into parent groups</span>
      <span class="n">parent_set_1</span> <span class="o">=</span> <span class="n">selected_indices</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
      <span class="n">parent_set_2</span> <span class="o">=</span> <span class="n">selected_indices</span><span class="p">[</span><span class="mi">1</span><span class="p">::</span><span class="mi">2</span><span class="p">]</span>
      <span class="c1"># Replicate each parent in their respective sets</span>
      <span class="n">parent_set_1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="n">parent_set_1</span><span class="p">,</span> <span class="n">offspring_per_parent</span><span class="p">)</span>
      <span class="n">parent_set_2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="n">parent_set_2</span><span class="p">,</span> <span class="n">offspring_per_parent</span><span class="p">)</span>
      <span class="c1"># Shuffle the parent sets based on mating strategy</span>
      <span class="k">if</span> <span class="ow">not</span> <span class="n">has_assortative_pairing</span><span class="p">:</span>
        <span class="n">rng</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">parent_set_2</span><span class="p">)</span>
      <span class="c1"># Combine the two parent sets to form genome pairs</span>
      <span class="n">parent_pairs_shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,)</span> <span class="o">+</span> <span class="n">env_target</span><span class="o">.</span><span class="n">shape</span> <span class="o">+</span> <span class="p">(</span><span class="mi">2</span><span class="p">,)</span>
      <span class="n">parent_pairs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">parent_pairs_shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
      <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="n">parent_set_1</span><span class="p">]</span>
      <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="n">parent_set_2</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"Invalid selection_type string, use one of proportional, softmax, proportional truncation, or deterministic truncation"</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">selection_type</span> <span class="o">!=</span> <span class="s1">'deterministic truncation'</span><span class="p">:</span>
      <span class="c1"># form parent pairs</span>
      <span class="n">parent_pairs_shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,)</span> <span class="o">+</span> <span class="n">env_target</span><span class="o">.</span><span class="n">shape</span> <span class="o">+</span> <span class="p">(</span><span class="mi">2</span><span class="p">,)</span>
      <span class="n">parent_pairs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">parent_pairs_shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>

      <span class="k">if</span> <span class="n">has_recombination</span><span class="p">:</span>
        <span class="n">selected_parent_indices_1</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">pop_size</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span>
                                               <span class="n">p</span><span class="o">=</span><span class="n">prob_scores</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">pair_directions</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span> <span class="o">-</span> <span class="mi">1</span>  <span class="c1"># Results in either -1 or 1</span>
        <span class="k">if</span> <span class="n">has_assortative_pairing</span><span class="p">:</span>
          <span class="c1"># Use pair direction for assortative mating</span>
          <span class="n">selected_parent_indices_2</span> <span class="o">=</span> <span class="n">selected_parent_indices_1</span> <span class="o">+</span> <span class="n">pair_directions</span>
        <span class="k">else</span><span class="p">:</span>
          <span class="c1"># For non-assortative mating, select a second set of parents</span>
          <span class="n">selected_parent_indices_2</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">pop_size</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span>
                                                 <span class="n">p</span><span class="o">=</span><span class="n">prob_scores</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
          <span class="c1"># Correct self-pairing</span>
          <span class="n">selected_parent_indices_2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span> <span class="c1">#ternary use of where</span>
            <span class="n">selected_parent_indices_1</span> <span class="o">==</span> <span class="n">selected_parent_indices_2</span><span class="p">,</span> <span class="c1"># if this condition</span>
            <span class="p">(</span><span class="n">selected_parent_indices_2</span> <span class="o">+</span> <span class="n">pair_directions</span><span class="p">)</span> <span class="o">%</span> <span class="n">pop_size</span><span class="p">,</span> <span class="c1"># do this</span>
            <span class="n">selected_parent_indices_2</span><span class="p">)</span> <span class="c1"># other wise do this</span>
        <span class="c1"># Adjust any out-of-bounds indices</span>
        <span class="n">selected_parent_indices_2</span><span class="p">[</span><span class="n">selected_parent_indices_2</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">selected_parent_indices_2</span><span class="p">[</span><span class="n">selected_parent_indices_2</span> <span class="o">&gt;=</span> <span class="n">pop_size</span><span class="p">]</span> <span class="o">=</span> <span class="n">pop_size</span> <span class="o">-</span> <span class="mi">1</span>
      <span class="k">else</span><span class="p">:</span> <span class="c1"># no recombination</span>
        <span class="n">selected_parent_indices_1</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">pop_size</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span>
                                               <span class="n">p</span><span class="o">=</span><span class="n">prob_scores</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">selected_parent_indices_2</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">pop_size</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span>
                                               <span class="n">p</span><span class="o">=</span><span class="n">prob_scores</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
      <span class="c1"># use the selected indices to form the parent genome pairs</span>
      <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="n">selected_parent_indices_1</span><span class="p">]</span>
      <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">=</span> <span class="n">population_genome</span><span class="p">[</span><span class="n">selected_parent_indices_2</span><span class="p">]</span>

    <span class="c1"># recombination</span>
    <span class="k">if</span> <span class="n">has_recombination</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">recombination_type</span> <span class="o">==</span> <span class="s1">'random'</span><span class="p">:</span>
        <span class="n">mask_shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,)</span> <span class="o">+</span> <span class="n">env_target</span><span class="o">.</span><span class="n">shape</span> <span class="o">+</span> <span class="p">(</span><span class="mi">2</span><span class="p">,)</span>
        <span class="n">mask1</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">mask_shape</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
        <span class="n">mask2</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">mask_shape</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">bool_</span><span class="p">)</span>
        <span class="n">children</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty_like</span><span class="p">(</span><span class="n">parent_pairs</span><span class="p">)</span>
        <span class="n">children</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask1</span><span class="p">,</span> <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span> <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">...</span><span class="p">])</span>
        <span class="n">children</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask2</span><span class="p">,</span> <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span> <span class="n">parent_pairs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">...</span><span class="p">])</span>
      <span class="k">elif</span> <span class="n">recombination_type</span> <span class="o">==</span> <span class="s1">'crossover'</span><span class="p">:</span>
        <span class="c1"># Flatten the genomes for crossover operations</span>
        <span class="c1"># Shape before flattening: (pop_size // 2, 2, num_blocks, block_size, 2)</span>
        <span class="n">flat_parent_pairs</span> <span class="o">=</span> <span class="n">parent_pairs</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
        <span class="c1"># Shape after flattening: (pop_size // 2, 2, num_blocks * block_size * 2)</span>
        <span class="c1"># Generate two sets of unique crossover points for each parent pair</span>
        <span class="n">crossover_points_1</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">flat_parent_pairs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,))</span>
        <span class="n">crossover_points_2</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">flat_parent_pairs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">pop_size</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,))</span>
        <span class="c1"># Create a range array that matches the last dimension of flat_parent_pairs</span>
        <span class="c1"># to help make the mask for crossover operations</span>
        <span class="n">range_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">flat_parent_pairs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1"># Use broadcasting to create masks: True if index is less than the crossover point</span>
        <span class="n">mask1</span> <span class="o">=</span> <span class="n">range_array</span> <span class="o">&lt;</span> <span class="n">crossover_points_1</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">]</span>
        <span class="n">mask2</span> <span class="o">=</span> <span class="n">range_array</span> <span class="o">&lt;</span> <span class="n">crossover_points_2</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">]</span>
        <span class="n">mask2</span> <span class="o">=</span> <span class="o">~</span><span class="n">mask2</span>
        <span class="n">children_flat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty_like</span><span class="p">(</span><span class="n">flat_parent_pairs</span><span class="p">)</span>
        <span class="n">children_flat</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask1</span><span class="p">,</span> <span class="n">flat_parent_pairs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">flat_parent_pairs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:])</span>
        <span class="n">children_flat</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mask2</span><span class="p">,</span> <span class="n">flat_parent_pairs</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">flat_parent_pairs</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:])</span>
        <span class="n">children</span> <span class="o">=</span> <span class="n">children_flat</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">parent_pairs</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
      <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"Invalid recombination_type string, use one of random or crossover"</span><span class="p">)</span>
      <span class="n">recombined_population_genome</span> <span class="o">=</span> <span class="n">children</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">genome_shape</span><span class="p">)</span>
      <span class="n">population_genome</span> <span class="o">=</span> <span class="n">recombined_population_genome</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="c1"># the population is just the selected parents</span>
      <span class="n">population_genome</span> <span class="o">=</span> <span class="n">parent_pairs</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">genome_shape</span><span class="p">)</span>

    <span class="c1"># mutation</span>
    <span class="k">if</span> <span class="n">has_mutation</span><span class="p">:</span>
      <span class="n">mutation_mask</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">population_genome</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">mutation_rate</span>
      <span class="n">population_genome</span> <span class="o">^=</span> <span class="n">mutation_mask</span>

    <span class="c1"># dynamic environment</span>
    <span class="k">if</span> <span class="n">has_changing_environment</span><span class="p">:</span>
      <span class="c1"># see which aspect of the environment change and update them</span>
      <span class="n">did_change</span> <span class="o">=</span> <span class="n">env_change_mask</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rng</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="n">env_change_mask</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">env_change_rate</span><span class="p">)</span>
      <span class="c1"># flip the bits where change happened</span>
      <span class="n">env_target</span><span class="p">[</span><span class="n">did_change</span><span class="p">]</span> <span class="o">=</span> <span class="o">~</span><span class="n">env_target</span><span class="p">[</span><span class="n">did_change</span><span class="p">]</span>

    <span class="c1"># repeat the loop for n_gen iterations</span>

  <span class="n">results</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">'mean_score_hist'</span><span class="p">:</span> <span class="n">mean_score_hist</span><span class="p">,</span>
    <span class="s1">'var_score_hist'</span><span class="p">:</span> <span class="n">var_score_hist</span><span class="p">,</span>
    <span class="s1">'skew_score_hist'</span><span class="p">:</span> <span class="n">skew_score_hist</span><span class="p">,</span>
    <span class="s1">'mean_good_bits_hist'</span><span class="p">:</span> <span class="n">mean_good_bits_hist</span>
  <span class="p">}</span>
  <span class="k">if</span> <span class="n">compute_IQR</span><span class="p">:</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">'lower_score_quartile_hist'</span><span class="p">]</span> <span class="o">=</span> <span class="n">lower_score_quartile_hist</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">'upper_score_quartile_hist'</span><span class="p">]</span> <span class="o">=</span> <span class="n">upper_score_quartile_hist</span><span class="p">,</span>
  <span class="k">if</span> <span class="n">has_changing_environment</span><span class="p">:</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">'mean_flexible_change_hist'</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean_flexible_change_hist</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">'mean_flexible_no_change_hist'</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean_flexible_no_change_hist</span>
  <span class="k">return</span> <span class="n">results</span>

<span class="c1"># Parameters</span>
<span class="n">common_params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">'n_gen'</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
    <span class="s1">'pop_size'</span><span class="p">:</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="s1">'init_pop'</span><span class="p">:</span> <span class="s1">'simple'</span><span class="p">,</span>
    <span class="s1">'seed'</span><span class="p">:</span> <span class="mi">123</span><span class="p">,</span>
    <span class="s1">'mutation_rate'</span><span class="p">:</span> <span class="mf">0.01</span><span class="p">,</span>
    <span class="s1">'has_mutation'</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">'has_recombination'</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">'has_assortative_pairing'</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">'recombination_type'</span><span class="p">:</span> <span class="s1">'crossover'</span><span class="p">,</span>
    <span class="s1">'has_learning'</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">'n_learning_trials'</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
    <span class="s1">'proportion_shift'</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span>
    <span class="s1">'proportion_scale'</span><span class="p">:</span> <span class="mf">19.0</span><span class="p">,</span>
    <span class="s1">'has_changing_environment'</span><span class="p">:</span><span class="kc">False</span><span class="p">,</span>
    <span class="s1">'change_env_prop'</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
    <span class="s1">'env_change_rate'</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
    <span class="s1">'compute_IQR'</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">}</span>

<span class="c1"># Create 3x1 subplots</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">))</span>

<span class="n">selection_types</span> <span class="o">=</span> <span class="p">[</span><span class="s1">'proportional'</span><span class="p">,</span> <span class="s1">'deterministic truncation'</span><span class="p">,</span> <span class="s1">'softmax'</span><span class="p">]</span>
<span class="n">base_colors</span> <span class="o">=</span> <span class="p">{</span><span class="s1">'proportional'</span><span class="p">:</span> <span class="s1">'Blues'</span><span class="p">,</span> <span class="s1">'deterministic truncation'</span><span class="p">:</span> <span class="s1">'Reds'</span><span class="p">,</span> <span class="s1">'softmax'</span><span class="p">:</span> <span class="s1">'Greens'</span><span class="p">}</span>
<span class="n">labels</span> <span class="o">=</span> <span class="p">{</span><span class="s1">'proportional'</span><span class="p">:</span> <span class="s1">'Proportional'</span><span class="p">,</span> <span class="s1">'deterministic truncation'</span><span class="p">:</span> <span class="s1">'Deterministic Truncation'</span><span class="p">,</span> <span class="s1">'softmax'</span><span class="p">:</span> <span class="s1">'Softmax'</span><span class="p">}</span>

<span class="c1"># Define parameter dictionaries for each block size</span>
<span class="c1">#vary num_blocks with block_size to keep constant genotype length = num_blocks * block_size</span>
<span class="n">block_sizes</span> <span class="o">=</span>       <span class="p">[</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">]</span>
<span class="n">numbers_of_blocks</span> <span class="o">=</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>  <span class="mi">2</span><span class="p">,</span>  <span class="mi">1</span><span class="p">]</span>
<span class="n">block_params</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">block_size</span><span class="p">,</span> <span class="n">num_blocks</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">block_sizes</span><span class="p">,</span> <span class="n">numbers_of_blocks</span><span class="p">):</span>
  <span class="n">params</span> <span class="o">=</span> <span class="p">{</span><span class="s1">'block_size'</span><span class="p">:</span> <span class="n">block_size</span><span class="p">,</span> <span class="s1">'num_blocks'</span><span class="p">:</span> <span class="n">num_blocks</span><span class="p">}</span>
  <span class="n">block_params</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>

<span class="c1"># Plot for each selection type</span>
<span class="k">for</span> <span class="n">ax</span><span class="p">,</span> <span class="n">selection_type</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">axs</span><span class="p">,</span> <span class="n">selection_types</span><span class="p">):</span>
  <span class="n">base_color</span> <span class="o">=</span> <span class="n">base_colors</span><span class="p">[</span><span class="n">selection_type</span><span class="p">]</span>
  <span class="n">base</span> <span class="o">=</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">colormaps</span><span class="p">[</span><span class="n">base_color</span><span class="p">]</span>
  <span class="n">num_shades</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">block_sizes</span><span class="p">)</span>
  <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">base</span><span class="p">((</span><span class="n">i</span><span class="o">+</span><span class="mi">2</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">num_shades</span><span class="o">+</span><span class="mi">2</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_shades</span><span class="p">)]</span>
  <span class="n">label</span> <span class="o">=</span> <span class="n">labels</span><span class="p">[</span><span class="n">selection_type</span><span class="p">]</span>

  <span class="k">for</span> <span class="n">block_param</span><span class="p">,</span> <span class="n">color</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">block_params</span><span class="p">,</span> <span class="n">colors</span><span class="p">):</span>
    <span class="c1"># Combine parameters</span>
    <span class="n">combined_params</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="n">common_params</span><span class="p">,</span>
                       <span class="o">**</span><span class="n">block_param</span><span class="p">,</span>
                       <span class="s1">'selection_type'</span><span class="p">:</span> <span class="n">selection_type</span><span class="p">}</span>
    <span class="n">r</span> <span class="o">=</span> <span class="n">evo_learning_simulation</span><span class="p">(</span><span class="o">**</span><span class="n">combined_params</span><span class="p">)</span>
    <span class="n">generations</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">combined_params</span><span class="p">[</span><span class="s1">'n_gen'</span><span class="p">])</span>

    <span class="c1"># normalized mean score and quartiles</span>
    <span class="n">mean_hist</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s1">'mean_score_hist'</span><span class="p">]</span><span class="o">/</span><span class="n">combined_params</span><span class="p">[</span><span class="s1">'num_blocks'</span><span class="p">]</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">generations</span><span class="p">,</span> <span class="n">mean_hist</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">label</span> <span class="o">+</span> <span class="s2">" - Block Size: </span><span class="si">{block_param['block_size']}</span><span class="s2">, Block Number: </span><span class="si">{block_param['num_blocks']}</span><span class="s2">"</span><span class="p">)</span>
  <span class="n">norm</span> <span class="o">=</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="n">block_sizes</span><span class="p">),</span> <span class="n">vmax</span><span class="o">=</span><span class="nb">max</span><span class="p">(</span><span class="n">block_sizes</span><span class="p">))</span>
  <span class="n">scalar_mappable</span> <span class="o">=</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">ScalarMappable</span><span class="p">(</span><span class="n">norm</span><span class="o">=</span><span class="n">norm</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">base</span><span class="p">)</span>

  <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Selection Type: </span><span class="si">{</span><span class="n">selection_type</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
  <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">'Generation'</span><span class="p">)</span>
  <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">'Normalized Mean Score'</span><span class="p">)</span>
  <span class="c1">#ax.legend()</span>
  <span class="n">cbar</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">scalar_mappable</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">orientation</span><span class="o">=</span><span class="s1">'vertical'</span><span class="p">)</span>
  <span class="n">cbar</span><span class="o">.</span><span class="n">set_label</span><span class="p">(</span><span class="s1">'Bits per Block'</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">remove_ip_clutter</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>

</div>
<p><a class="reference external" href="https://github.com/dcownden/PerennialProblemsOfLifeWithABrain/tree/main//sequences/P1C3_RealEvolution/solutions/P1C3_Sequence4_Solution_9401b50a.py"><em>Click for solution</em></a></p>
<p><em>Example output:</em></p>
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/dcownden/PerennialProblemsOfLifeWithABrain/main/sequences/P1C3_RealEvolution/static/P1C3_Sequence4_Solution_9401b50a_0.png"><img align="center" alt="Solution hint" class="align-center" src="https://raw.githubusercontent.com/dcownden/PerennialProblemsOfLifeWithABrain/main/sequences/P1C3_RealEvolution/static/P1C3_Sequence4_Solution_9401b50a_0.png" style="width: 1000.0px; height: 1500.0px;"/></a>
<p>In the simulations above we see the evolution of the (normalized) mean fitness score for different selection types, and different numbers of bits required to align before a fitness increment ([2, 5, 10, 20]). In each simulation there is mutation (but no recombination) to introduce new variance, yet the evolutionary process is unable to hit upon having all 20 bits in a receptor align at once (darkest lines) over 500 generations for any of these selection types. Fitness score functions like this, where a large co-adaption is required for any fitness increment, are sometimes referred to as ‘needle-in-a-haystack’ as there is one correct configuration, the needle, hidden among many, in this case roughly one million (<span class="math notranslate nohighlight">\(2^{20}\)</span>), incorrect configurations, the haystack. What makes this such a challenging problem for an evolutionary process is that all of the incorrect configurations look the same to selection; even though a configuration of one ‘wrong’ bit and nineteen ‘right’ bits is very close to the correct configuration in the genome space, from the perspective of selection, this close but wrong configuration is indistinguishable from any of the other wrong configurations. Selection gives no ‘hot’ or ‘cold’ style hints in this case to guide the evolutionary search. This is in stark contrast to the similar problem we looked at in the beginning of the sequence where selection was sensitive to every changing bit in the genome. For subsequent simulations we use ‘proportional’ selection, primarily for the sake of following Hinton and Nolan 1987.</p>
<p>Note though that even if the high scoring genotype is discovered through mutation and recombination, selection may not always be able to grow the frequency of this genotype within the population. Search and discovery with a so called ‘flat’ fitness landscape is just one aspect of the issue here. Additionally when using evolutionary search to find a large, tightly co-adapted configuration, recombination can be a liability. Consider that when this co-adaptation is discovered, if there is recombination then it is not easily passed on to descendants. When an adapted individual mates with any individual other than one almost identical to itself, the co-adaptation is likely to be destroyed by recombination. The crux of the issue is that for large co-adaptions very few, if any, improved genotypes are likely to appear in a given generation, and unless mating is strongly assortative these individuals are unlikely to mate with each other. For the co-adaptation to persist from generation to generation each good genotype must, on average, give rise to at least one good offspring in the next generation. If the dispersal of complex co-adaptations due to mating causes each good genotype to have less than one expected good offspring, the co-adaptation will not spread, even if it is discovered many times.</p>
</div>
<div class="section" id="math-exercise-probabilities-of-evolutionary-events">
<h2>Math Exercise: Probabilities of Evolutionary Events.<a class="headerlink" href="#math-exercise-probabilities-of-evolutionary-events" title="Permalink to this headline">¶</a></h2>
<p>With some combination of large population size and many generations, mutation alone (without recombination) will eventually hit upon this rare configuration. Assume a population of size <span class="math notranslate nohighlight">\(N\)</span> where each individual has a single block of <span class="math notranslate nohighlight">\(n\)</span> bits to be aligned.</p>
<p><strong>Questions:</strong></p>
<ol class="simple">
<li><p>Calculate the approximate probability of an offspring with a ‘good block’ being born in a population where none of the previous generations have had a ‘good block’. Use the following simplifying assumptions and hint:</p></li>
</ol>
<blockquote>
<div><ul class="simple">
<li><p>Each bit in a randomly selected parent is equally likely to be a 1 or a 0. This will be close to true for large populations with some mutation and no selective pressure.</p></li>
<li><p>Mutation rate and genome size are such that 0.2 of the population has a single mutation in a given generation, and double mutations within an individual are rare enough to be ignored. (As is roughly the case in a genome of 20 bits with a per bit mutation rate of 0.01.)</p></li>
<li><p>Hint: Use the first order Taylor series approximation for <span class="math notranslate nohighlight">\((1-p)^x\)</span>, around <span class="math notranslate nohighlight">\(p=0\)</span>.</p></li>
</ul>
</div></blockquote>
<ol class="simple">
<li><p>If the probability of at least one ‘genetic lottery winner’ in a given generation is <span class="math notranslate nohighlight">\(q\)</span>, how many generations on average will it take for the evolutionary process to have such a winner. Hint: Look up the Geometric distribution on Wikipedia.</p></li>
<li><p>Now, let’s consider the scenario with recombination. How does this affect the time taken to find a co-adapted block?</p></li>
<li><p>How many generations on average would it take to get at least two co-adapted individuals in one generation (which would then mate and spread the adaptation), making the same assumptions as before about no adapted parents in the previous generations.</p></li>
</ol>
<p><strong>Answers:</strong></p>
<ol class="simple">
<li><p>The probability of any individual being a ‘genetic lottery’ winner is <span class="math notranslate nohighlight">\(p= 0.2 \cdot \frac{1}{2^n}\)</span> representing the chance of a parent having all but one bit aligned and then mutation aligning the last bit in their offspring. With <span class="math notranslate nohighlight">\(N\)</span> individuals per generation, the probability of no genetic lottery winners is <span class="math notranslate nohighlight">\((1-p)^N\)</span>. Using a first-order Taylor series approximation around <span class="math notranslate nohighlight">\(p=0\)</span>, we find that the chance of at least one winner in a generation is <span class="math notranslate nohighlight">\(q \approx 1-(1-Np) = \frac{N}{5 \cdot2^n}\)</span>.</p></li>
<li><p>For a geometrically distributed event like this, where <span class="math notranslate nohighlight">\(q\)</span> is the success probability, the average waiting time is the reciprocal of <span class="math notranslate nohighlight">\(q\)</span>. Therefore, we expect to wait about <span class="math notranslate nohighlight">\(\frac{5 \cdot2^n}{N}\)</span> generations for a ‘genetic lottery winner’. Note though that this wait time is highly variable with a standard deviation roughly equal to the mean for small <span class="math notranslate nohighlight">\(q\)</span>. The number of generations required grows exponentially in the number of bits that need to be correct, but only decrease like one over the number of individuals, so increasing population size is not a scalable solution.</p></li>
<li><p>With recombination, every offspring, not just those with mutations, has a chance of being a ‘genetic lottery’ winner. While this increases the likelihood of finding a co-adapted block, there’s also a higher chance it’ll be lost in subsequent generations due to recombination. Thus, the average wait time decreases to <span class="math notranslate nohighlight">\(\frac{2^n}{N}\)</span>.</p></li>
<li><p>To find the average number of generations needed for at least two ‘genetic lottery winners’, we calculate the probability of this event occurring in a single generation. This calculation considers the likelihood of having zero winners and exactly one winner. Then, we use the Taylor series expansion to simplify our expression, arriving at an approximate average wait time of <span class="math notranslate nohighlight">\(\frac{2^{2n}}{N^2}\)</span> generations. Specifically, with recombination the probability that any individual is a ‘genetic lottery’ winner is the <span class="math notranslate nohighlight">\(p= \frac{1}{2^n}\)</span>. The probability of zero such winners is <span class="math notranslate nohighlight">\((1-p)^N\)</span> and the probability of exactly one such winners is <span class="math notranslate nohighlight">\(N p (1-p)^{N-1}\)</span>. Then using the same Taylor expansion as before and noting that <span class="math notranslate nohighlight">\((1-p)^N \approx (1-p)^{N-1}\)</span> for large <span class="math notranslate nohighlight">\(N\)</span> and small <span class="math notranslate nohighlight">\(p\)</span>, we have that the probability of there not being at least two ‘genetic lottery winners’ in a given generation is <span class="math notranslate nohighlight">\(q = 1 - (1-p)^N - Np(1-p)^{N-1} \approx 1 - (1-Np) - Np(1-Np) = (Np)^2\)</span>. So on average is will take <span class="math notranslate nohighlight">\(\frac{2^{2n}}{N^2}\)</span> generations. We could be waiting a long time. Note though that this will be a slight over-estimate of the waiting time, as ‘singleton’ lottery winners will emerge occasionally, and over time they will seed the population with many similar and nearly adapted phenotypes, so that a) the probability of ‘double winners’ increases, and b) the probability of a ‘singleton winner’ having adapted offspring because they mated with a sufficiently similar non-adapted individual increase.</p></li>
</ol>
<p>Submit your feedback</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># @markdown Submit your feedback</span>
<span class="n">content_review</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">feedback_prefix</span><span class="si">}</span><span class="s2">_M2"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/vnd.jupyter.widget-view+json">
{"version_major": 2, "version_minor": 0, "model_id": "ce224024c3cc436c8cb33e0c3cae6905"}
</script></div>
</div>
</div>
</div>
<div class="section" id="learning-can-accelerate-adaptive-response">
<h1>1.3.4.3 Learning Can Accelerate Adaptive Response<a class="headerlink" href="#learning-can-accelerate-adaptive-response" title="Permalink to this headline">¶</a></h1>
<p>From the calculations above we can see that increasing population size isn’t really a viable option for accelerating evolutionary discovery of ‘good’ parameters for ‘difficult’ fitness functions. In the first part of this sequence we saw how learning can be an adaptive response to a rapidly changing environment. We will now use our simulation to explore how incorporating learning within an individual’s lifetime can significantly accelerate the evolutionary process, guiding populations more rapidly towards optimal parameters. We run the same simulations as above, but now we allow for 1000 learning trials within each individuals lifetime.</p>
<div class="section" id="todo-pciture-evolution-and-learning-as-inner-and-outer-loops-of-one-optimization-process">
<h2><strong>TODO</strong> Pciture: Evolution and Learning as Inner and Outer Loops of One Optimization Process.<a class="headerlink" href="#todo-pciture-evolution-and-learning-as-inner-and-outer-loops-of-one-optimization-process" title="Permalink to this headline">¶</a></h2>
<p>Within-lifetime learning as the inner loop, between generation evolutionary adaptation as the inner loop.</p>
<p><strong>Run this cell</strong> to see how learning and recombination interact when a population is adapting to a ‘needle-in-a-haystack’ fitness landscape.</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># @markdown **Run this cell** to see how learning and recombination interact when a population is adapting to a 'needle-in-a-haystack' fitness landscape.</span>

<span class="n">common_params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">'selection_type'</span><span class="p">:</span> <span class="s1">'proportional'</span><span class="p">,</span>
    <span class="s1">'softmax_temp'</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
    <span class="s1">'seed'</span><span class="p">:</span> <span class="mi">234</span><span class="p">,</span>
    <span class="s1">'pop_size'</span><span class="p">:</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="s1">'num_blocks'</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>     <span class="c1"># Focus on 1 trait</span>
    <span class="s1">'block_size'</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span>  <span class="c1"># Using 20 bits per block</span>
    <span class="s1">'has_mutation'</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">'mutation_rate'</span><span class="p">:</span> <span class="mf">0.01</span><span class="p">,</span>
    <span class="s1">'has_recombination'</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">'has_assortative_pairing'</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">'recombination_type'</span><span class="p">:</span> <span class="s1">'random'</span><span class="p">,</span>
    <span class="s1">'proportion_shift'</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span>
    <span class="s1">'proportion_scale'</span><span class="p">:</span> <span class="mf">19.0</span><span class="p">,</span>
    <span class="s1">'has_changing_environment'</span><span class="p">:</span><span class="kc">False</span><span class="p">,</span>
    <span class="s1">'change_env_prop'</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
    <span class="s1">'env_change_rate'</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
    <span class="s1">'compute_IQR'</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">'init_pop'</span><span class="p">:</span> <span class="s1">'simple'</span>
<span class="p">}</span>

<span class="c1"># Parameters specific to the learning scenario</span>
<span class="n">learning_params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">'has_learning'</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">'n_learning_trials'</span><span class="p">:</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="s1">'n_gen'</span><span class="p">:</span> <span class="mi">50</span><span class="p">,</span>
<span class="p">}</span>

<span class="c1"># Parameters specific to the scenario without learning</span>
<span class="n">no_learning_params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">'has_learning'</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">'n_learning_trials'</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>  <span class="c1"># Learning is not occurring</span>
    <span class="s1">'n_gen'</span><span class="p">:</span> <span class="mi">10000</span> <span class="c1">#</span>
<span class="p">}</span>

<span class="n">recombination_params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">'has_recombination'</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">}</span>

<span class="n">no_recombination_params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">'has_recombination'</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">}</span>

<span class="c1"># Create 2x1 subplots for comparison</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

<span class="c1"># Function to plot results</span>
<span class="k">def</span> <span class="nf">plot_results</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">simulation_params</span><span class="p">,</span> <span class="n">title</span><span class="p">,</span> <span class="n">label</span><span class="p">):</span>
  <span class="n">results</span> <span class="o">=</span> <span class="n">evo_learning_simulation</span><span class="p">(</span><span class="o">**</span><span class="n">simulation_params</span><span class="p">)</span>
  <span class="n">generations</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">simulation_params</span><span class="p">[</span><span class="s1">'n_gen'</span><span class="p">])</span>
  <span class="c1"># Calculate normalized mean score</span>
  <span class="n">mean_hist</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">'mean_score_hist'</span><span class="p">]</span> <span class="o">/</span> <span class="n">simulation_params</span><span class="p">[</span><span class="s1">'num_blocks'</span><span class="p">]</span>
  <span class="c1"># Plot</span>
  <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">generations</span><span class="p">,</span> <span class="n">mean_hist</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">'blue'</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">label</span><span class="p">)</span>
  <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
  <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">'Generation'</span><span class="p">)</span>
  <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">'Normalized Mean Score'</span><span class="p">)</span>

<span class="c1"># Modify the plot_results function call for each subplot</span>
<span class="n">plot_results</span><span class="p">(</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">{</span><span class="o">**</span><span class="n">common_params</span><span class="p">,</span> <span class="o">**</span><span class="n">no_learning_params</span><span class="p">,</span> <span class="o">**</span><span class="n">no_recombination_params</span><span class="p">},</span> <span class="s2">"Proportional Selection</span><span class="se">\n</span><span class="s2">No Learning, No Recombination"</span><span class="p">,</span> <span class="s1">'No Learn/No Recomb'</span><span class="p">)</span>
<span class="n">plot_results</span><span class="p">(</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">{</span><span class="o">**</span><span class="n">common_params</span><span class="p">,</span> <span class="o">**</span><span class="n">no_learning_params</span><span class="p">,</span> <span class="o">**</span><span class="n">recombination_params</span><span class="p">},</span> <span class="s2">"Proportional Selection</span><span class="se">\n</span><span class="s2">No Learning, With Recombination"</span><span class="p">,</span> <span class="s1">'No Learn/Recomb'</span><span class="p">)</span>
<span class="n">plot_results</span><span class="p">(</span><span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">{</span><span class="o">**</span><span class="n">common_params</span><span class="p">,</span> <span class="o">**</span><span class="n">learning_params</span><span class="p">,</span> <span class="o">**</span><span class="n">no_recombination_params</span><span class="p">},</span> <span class="s2">"Proportional Selection</span><span class="se">\n</span><span class="s2">With Learning, No Recombination"</span><span class="p">,</span> <span class="s1">'Learn/No Recomb'</span><span class="p">)</span>
<span class="n">plot_results</span><span class="p">(</span><span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">{</span><span class="o">**</span><span class="n">common_params</span><span class="p">,</span> <span class="o">**</span><span class="n">learning_params</span><span class="p">,</span> <span class="o">**</span><span class="n">recombination_params</span><span class="p">},</span> <span class="s2">"Proportional Selection</span><span class="se">\n</span><span class="s2">With Learning and Recombination"</span><span class="p">,</span> <span class="s1">'Learn/Recomb'</span><span class="p">)</span>


<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="c1"># Assuming remove_ip_clutter is a predefined function for cleaning up the plots</span>
<span class="n">remove_ip_clutter</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>

</div>
<p>In the simulation shown in the top right, which features recombination without learning, we observe the emergence of singleton co-adaptations (as indicated by small spikes in the population’s mean fitness score). This aligns with our earlier mathematical predictions, yet these co-adaptations fail to spread across the population. Using the approximation from question 4 in the previous section, for a population with 1000 individuals trying to hit upon a 20-bit co-adaptation, the average waiting time until there are two adapted individuals in a single generation is roughly a million generations. If you have time on your hands you could adapt the parameters of the simulation and see how long you have to wait!</p>
<p>In all the other cases the co-adaptation is discovered relatively quickly. Without learning (or recombination, top left), in this one simulation, the co-adaptation is discovered after ~6000 generations of evolutionary grinding. From exercise 2 above, this is roughly aligned with the geometrically distributed waiting time with mean of approximately 5000. (Though several simulations are required to check this alignment more fully.)</p>
<p>In the cases with learning (bottom), the co-adaptation is discovered and spreads within tens, not thousands of generations. This takes slightly longer with recombination (bottom right), but not insurmountably so. Now comparing generational times directly is a bit misleading. In each generation with learning, (in these simulations) each individual could make up to 1000 distinct function evaluations, though given their learning rules they often made far fewer (distinct) evaluations. In stark contrast, in the simulations without learning each individual in each generation made exactly one function evaluation. So from a strict number of function evaluations perspective learning isn’t really helping us directly, at least not in the simulations with no recombination.</p>
<p>Our earlier simulations and mathematical models in section 1.3.2 demonstrated that recombination significantly accelerates evolutionary search in simpler fitness landscapes. However, in the context of more complex fitness score functions as explored here, recombination tends to be less advantageous, potentially even a liability, although learning can mitigate this to an extent. It seems as though the ideal situation would be to have a recombination structured in such a way that it both accelerated evolutionary search in some situations, without being an ‘adaptive liability’ in others. Indeed, this is precisely what we see in the natural world. In recombination, chromosomes are not shattered and shuffled and thrown to the wind all willy nilly, instead, specific break-points where cross-over style recombination events are more likely exist and are situated so as to preserve groups of tightly co-adapted genes. We will not explore this particular idea with simulations, but we can imagine how the genetic mechanisms for setting and maintaining good ‘break-points’ is itself the result of evolution. Aspects of life history and genetic replication that serve as the foundational assumptions of some evolutionary models, can themselves be understood as the results of evolution by natural selection. They key thing to keep in mind is that evolution is a ‘black-box’ optimizer, it has and uses no knowledge of the underlying structure of the fitness function, selection acts solely on the output of this function. (Or in some very real sense selection <strong>is</strong> the evaluation of the fitness function.) And so we expect all viable ‘adaptive accelerators’ to be deployed against ‘the vague uncertainties of life’.</p>
</div>
<div class="section" id="todo-bio-box-facilitated-variation">
<h2><strong>TODO</strong> Bio Box: Facilitated Variation<a class="headerlink" href="#todo-bio-box-facilitated-variation" title="Permalink to this headline">¶</a></h2>
<p>We conclude this section on learning and evolution by noting that we used a neural circuit as our motivating sketch of a highly co-adapted trait, and how this relates to lifespan. In an evolutionary context, a longer lifespan not only offers extended periods for learning and adaptation but also allows an organism to maximize the benefits of its investment in learning. This investment isn’t trivial—it encompasses both the time devoted to learning and the sophisticated neural architecture required to support it. Our simulations further highlight that complex behaviors, which often involve numerous genetic co-adaptations, are challenging for evolution to both discover and spread within a population. Consequently, we expect, and generally observe, a confluence of extended lifespans, advanced learning capabilities, and complex behaviors in various life forms. This is because each of these traits supports and enhances the others, creating a synergy that is advantageous for survival and reproduction in diverse and changing environments. Brainy, motor-flexible animals, with rich senses exemplify this, and allow us to think about what brains are for in an evolutionary sense: learning adaptive behaviours quickly.</p>
<p>Submit your feedback</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># @markdown Submit your feedback</span>
<span class="n">content_review</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">feedback_prefix</span><span class="si">}</span><span class="s2">_M3"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/vnd.jupyter.widget-view+json">
{"version_major": 2, "version_minor": 0, "model_id": "0e56ba5701c3409b9e6e0cf6657fef96"}
</script></div>
</div>
</div>
</div>
<script type="application/vnd.jupyter.widget-state+json">
{"state": {"762676c9a2774b5abf3e73a8393b2b6c": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": "0.5em", "right": null, "top": null, "visibility": null, "width": "auto"}}, "ef10cc7fbacb46c5bdd6449cb1825d96": {"model_name": "ButtonStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "button_color": "#aaffaa", "font_weight": ""}}, "681145ce54324294bf48fc5fa4a562e8": {"model_name": "ButtonModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": ["happy"], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "ButtonView", "button_style": "", "description": "\ud83d\ude42", "disabled": false, "icon": "", "layout": "IPY_MODEL_762676c9a2774b5abf3e73a8393b2b6c", "style": "IPY_MODEL_ef10cc7fbacb46c5bdd6449cb1825d96", "tooltip": "happy"}}, "86fec608b60f4c948d8a4db30edfa707": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": "0.5em", "right": null, "top": null, "visibility": null, "width": "auto"}}, "1e56082693984089bee919eea9dd4821": {"model_name": "ButtonStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "button_color": "#dddd77", "font_weight": ""}}, "c0008b9756ad48bd98d4c3a2657c5c0b": {"model_name": "ButtonModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": ["medium"], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "ButtonView", "button_style": "", "description": "\ud83d\ude10", "disabled": false, "icon": "", "layout": "IPY_MODEL_86fec608b60f4c948d8a4db30edfa707", "style": "IPY_MODEL_1e56082693984089bee919eea9dd4821", "tooltip": "medium"}}, "be66bd33711742829563e723894dd5fa": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": "0.5em", "right": null, "top": null, "visibility": null, "width": "auto"}}, "3bfd28a47649487fbea1a7a08ee2bbe9": {"model_name": "ButtonStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "button_color": "#ffaaaa", "font_weight": ""}}, "0ca2c9839cca4cf29a644d292da669b6": {"model_name": "ButtonModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": ["sad"], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "ButtonView", "button_style": "", "description": "\ud83d\ude41", "disabled": false, "icon": "", "layout": "IPY_MODEL_be66bd33711742829563e723894dd5fa", "style": "IPY_MODEL_3bfd28a47649487fbea1a7a08ee2bbe9", "tooltip": "sad"}}, "9d08cc76c60f46fbba37a137a1e24bfe": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": "auto"}}, "499b8d12394140b4a6db46d646e95cdc": {"model_name": "DescriptionStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "DescriptionStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "description_width": ""}}, "c3874a6600754d9c9e7b68865fcf79d4": {"model_name": "TextareaModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "TextareaModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "TextareaView", "continuous_update": true, "description": "", "description_tooltip": null, "disabled": false, "layout": "IPY_MODEL_9d08cc76c60f46fbba37a137a1e24bfe", "placeholder": "We want your feedback!", "rows": null, "style": "IPY_MODEL_499b8d12394140b4a6db46d646e95cdc", "value": ""}}, "c5d6696e6c3245b293ffae89ba43eaed": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": "auto"}}, "38c4830309ff4b41968b7ed466970d72": {"model_name": "ButtonStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "button_color": null, "font_weight": ""}}, "289fabf6b339456f91d3aceabdd90e40": {"model_name": "ButtonModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "ButtonView", "button_style": "", "description": "Submit", "disabled": false, "icon": "", "layout": "IPY_MODEL_c5d6696e6c3245b293ffae89ba43eaed", "style": "IPY_MODEL_38c4830309ff4b41968b7ed466970d72", "tooltip": ""}}, "aeae7449174345f49458cc018788b25a": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": "none", "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "18624d06211f420b96a5b87b1deef1c0": {"model_name": "HBoxModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "HBoxModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "HBoxView", "box_style": "", "children": ["IPY_MODEL_c3874a6600754d9c9e7b68865fcf79d4", "IPY_MODEL_289fabf6b339456f91d3aceabdd90e40"], "layout": "IPY_MODEL_aeae7449174345f49458cc018788b25a"}}, "16f53ddc8bdd40ffa4c7615fe4b50d53": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": "none", "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "d25c26cec9ae46e29b77abe96bb6d602": {"model_name": "DescriptionStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "DescriptionStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "description_width": ""}}, "ab53b6912d4e4897b6983bf6d3e37a46": {"model_name": "LabelModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "LabelModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "LabelView", "description": "", "description_tooltip": null, "layout": "IPY_MODEL_16f53ddc8bdd40ffa4c7615fe4b50d53", "placeholder": "\u200b", "style": "IPY_MODEL_d25c26cec9ae46e29b77abe96bb6d602", "value": "Thanks for your feedback!"}}, "9b12b1e3cf7f40cebfa305426851e220": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "c7cbd08e25b8441baec80760843c3f55": {"model_name": "HBoxModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "HBoxModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "HBoxView", "box_style": "", "children": ["IPY_MODEL_681145ce54324294bf48fc5fa4a562e8", "IPY_MODEL_c0008b9756ad48bd98d4c3a2657c5c0b", "IPY_MODEL_0ca2c9839cca4cf29a644d292da669b6"], "layout": "IPY_MODEL_9b12b1e3cf7f40cebfa305426851e220"}}, "fbe79da93d8145bfa045bcc9642f801f": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "ad857f49d2ac47989600d8f495735286": {"model_name": "VBoxModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "VBoxModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "VBoxView", "box_style": "", "children": ["IPY_MODEL_c7cbd08e25b8441baec80760843c3f55", "IPY_MODEL_18624d06211f420b96a5b87b1deef1c0", "IPY_MODEL_ab53b6912d4e4897b6983bf6d3e37a46"], "layout": "IPY_MODEL_fbe79da93d8145bfa045bcc9642f801f"}}, "eee5ed93425b4db08f2c44b899495f29": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "4bacb49152ae4176b4715917629a73c0": {"model_name": "VBoxModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "VBoxModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "VBoxView", "box_style": "", "children": ["IPY_MODEL_ad857f49d2ac47989600d8f495735286"], "layout": "IPY_MODEL_eee5ed93425b4db08f2c44b899495f29"}}, "266dcea6effc4f0a8032108c2c1a399b": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": "0.5em", "right": null, "top": null, "visibility": null, "width": "auto"}}, "d393e5227bb949d39d093d72b253c56a": {"model_name": "ButtonStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "button_color": "#aaffaa", "font_weight": ""}}, "7b069d8f3f0249ed8dc1f8ec29c67b19": {"model_name": "ButtonModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": ["happy"], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "ButtonView", "button_style": "", "description": "\ud83d\ude42", "disabled": false, "icon": "", "layout": "IPY_MODEL_266dcea6effc4f0a8032108c2c1a399b", "style": "IPY_MODEL_d393e5227bb949d39d093d72b253c56a", "tooltip": "happy"}}, "c8ac1b96af084b0a92137c074b32f108": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": "0.5em", "right": null, "top": null, "visibility": null, "width": "auto"}}, "e4e5be68f95c494a8026e4489dbfbc42": {"model_name": "ButtonStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "button_color": "#dddd77", "font_weight": ""}}, "8f247cddbff94860be3a2ff94c054e5a": {"model_name": "ButtonModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": ["medium"], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "ButtonView", "button_style": "", "description": "\ud83d\ude10", "disabled": false, "icon": "", "layout": "IPY_MODEL_c8ac1b96af084b0a92137c074b32f108", "style": "IPY_MODEL_e4e5be68f95c494a8026e4489dbfbc42", "tooltip": "medium"}}, "c5fe97d0e49f45fabb9c83b3ab1d03e9": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": "0.5em", "right": null, "top": null, "visibility": null, "width": "auto"}}, "1574b15d20de4a5fa2a4368bf3ddc665": {"model_name": "ButtonStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "button_color": "#ffaaaa", "font_weight": ""}}, "b083ab3fe9b54fa999d2eb8d1f5a7007": {"model_name": "ButtonModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": ["sad"], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "ButtonView", "button_style": "", "description": "\ud83d\ude41", "disabled": false, "icon": "", "layout": "IPY_MODEL_c5fe97d0e49f45fabb9c83b3ab1d03e9", "style": "IPY_MODEL_1574b15d20de4a5fa2a4368bf3ddc665", "tooltip": "sad"}}, "d1a5bd5648ed494c9b59e5f0cfd7c79a": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": "auto"}}, "31c1f0bb59f348a8ac3659affa4c20f8": {"model_name": "DescriptionStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "DescriptionStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "description_width": ""}}, "64ced597567e45b1a17ee1224064ba26": {"model_name": "TextareaModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "TextareaModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "TextareaView", "continuous_update": true, "description": "", "description_tooltip": null, "disabled": false, "layout": "IPY_MODEL_d1a5bd5648ed494c9b59e5f0cfd7c79a", "placeholder": "We want your feedback!", "rows": null, "style": "IPY_MODEL_31c1f0bb59f348a8ac3659affa4c20f8", "value": ""}}, "8d9aa329dd1949c28628b333b42669a3": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": "auto"}}, "b57abc8beba54ab392b595dabde56e09": {"model_name": "ButtonStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "button_color": null, "font_weight": ""}}, "fd3319403a604513871954d96571bb56": {"model_name": "ButtonModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "ButtonView", "button_style": "", "description": "Submit", "disabled": false, "icon": "", "layout": "IPY_MODEL_8d9aa329dd1949c28628b333b42669a3", "style": "IPY_MODEL_b57abc8beba54ab392b595dabde56e09", "tooltip": ""}}, "62ef04b7ae2d40128517ce92a1f68a32": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": "none", "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "469528debe4f45a2bdb6f7c8818c2a67": {"model_name": "HBoxModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "HBoxModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "HBoxView", "box_style": "", "children": ["IPY_MODEL_64ced597567e45b1a17ee1224064ba26", "IPY_MODEL_fd3319403a604513871954d96571bb56"], "layout": "IPY_MODEL_62ef04b7ae2d40128517ce92a1f68a32"}}, "3b78bc3a785e40fe88fbf8e38d1937a8": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": "none", "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "daa8105af0aa4287933c6a79325d7c38": {"model_name": "DescriptionStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "DescriptionStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "description_width": ""}}, "ce573385e5544cc4b2ac8e704862f5e6": {"model_name": "LabelModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "LabelModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "LabelView", "description": "", "description_tooltip": null, "layout": "IPY_MODEL_3b78bc3a785e40fe88fbf8e38d1937a8", "placeholder": "\u200b", "style": "IPY_MODEL_daa8105af0aa4287933c6a79325d7c38", "value": "Thanks for your feedback!"}}, "5f5e66585c784567bd984a11b1071b0e": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "35a8382c8e6f46c19a5a32393c638837": {"model_name": "HBoxModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "HBoxModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "HBoxView", "box_style": "", "children": ["IPY_MODEL_7b069d8f3f0249ed8dc1f8ec29c67b19", "IPY_MODEL_8f247cddbff94860be3a2ff94c054e5a", "IPY_MODEL_b083ab3fe9b54fa999d2eb8d1f5a7007"], "layout": "IPY_MODEL_5f5e66585c784567bd984a11b1071b0e"}}, "ea762df88a1a4f5d93b4b667c64dd30a": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "ca71862170b64767b177ca8dad1fdaca": {"model_name": "VBoxModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "VBoxModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "VBoxView", "box_style": "", "children": ["IPY_MODEL_35a8382c8e6f46c19a5a32393c638837", "IPY_MODEL_469528debe4f45a2bdb6f7c8818c2a67", "IPY_MODEL_ce573385e5544cc4b2ac8e704862f5e6"], "layout": "IPY_MODEL_ea762df88a1a4f5d93b4b667c64dd30a"}}, "14c85b569cd14556a87b47572e38236b": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "ce224024c3cc436c8cb33e0c3cae6905": {"model_name": "VBoxModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "VBoxModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "VBoxView", "box_style": "", "children": ["IPY_MODEL_ca71862170b64767b177ca8dad1fdaca"], "layout": "IPY_MODEL_14c85b569cd14556a87b47572e38236b"}}, "fcd57b782b8a4fdca4184903ae64510b": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "4296cc1046074ca5b5ecc01a1018087f": {"model_name": "MPLCanvasModel", "model_module": "jupyter-matplotlib", "model_module_version": "^0.11", "state": {"_cursor": "pointer", "_data_url": null, "_dom_classes": [], "_figure_label": "Figure", "_image_mode": "full", "_message": "", "_model_module": "jupyter-matplotlib", "_model_module_version": "^0.11", "_model_name": "MPLCanvasModel", "_rubberband_height": 0, "_rubberband_width": 0, "_rubberband_x": 0, "_rubberband_y": 0, "_size": [0, 0], "_view_count": null, "_view_module": "jupyter-matplotlib", "_view_module_version": "^0.11", "_view_name": "MPLCanvasView", "capture_scroll": false, "footer_visible": true, "header_visible": true, "layout": "IPY_MODEL_fcd57b782b8a4fdca4184903ae64510b", "pan_zoom_throttle": 33.0, "resizable": true, "toolbar": "IPY_MODEL_8823e2d491d941a4b7b6544f7c5a8e71", "toolbar_position": "left", "toolbar_visible": "fade-in-fade-out"}}, "9c35a3e701ad41dba2caba662e3ee9b2": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "8823e2d491d941a4b7b6544f7c5a8e71": {"model_name": "ToolbarModel", "model_module": "jupyter-matplotlib", "model_module_version": "^0.11", "state": {"_current_action": "", "_dom_classes": [], "_model_module": "jupyter-matplotlib", "_model_module_version": "^0.11", "_model_name": "ToolbarModel", "_view_count": null, "_view_module": "jupyter-matplotlib", "_view_module_version": "^0.11", "_view_name": "ToolbarView", "button_style": "", "collapsed": true, "layout": "IPY_MODEL_9c35a3e701ad41dba2caba662e3ee9b2", "orientation": "vertical", "toolitems": [["Home", "Reset original view", "home", "home"], ["Back", "Back to previous view", "arrow-left", "back"], ["Forward", "Forward to next view", "arrow-right", "forward"], ["Pan", "Left button pans, Right button zooms\nx/y fixes axis, CTRL fixes aspect", "arrows", "pan"], ["Zoom", "Zoom to rectangle\nx/y fixes axis", "square-o", "zoom"], ["Download", "Download plot", "floppy-o", "save_figure"]]}}, "ac085e6f440949afafd544dc66981884": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": "0.5em", "right": null, "top": null, "visibility": null, "width": "auto"}}, "b5ab8fe24f10491999e06a2c586aef17": {"model_name": "ButtonStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "button_color": "#aaffaa", "font_weight": ""}}, "23516cddded64f60ad1e8eb82fc32585": {"model_name": "ButtonModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": ["happy"], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "ButtonView", "button_style": "", "description": "\ud83d\ude42", "disabled": false, "icon": "", "layout": "IPY_MODEL_ac085e6f440949afafd544dc66981884", "style": "IPY_MODEL_b5ab8fe24f10491999e06a2c586aef17", "tooltip": "happy"}}, "d3238d0d8b78412a9f540b7ae05c5c08": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": "0.5em", "right": null, "top": null, "visibility": null, "width": "auto"}}, "1c70e7abfdfd4d17a8cee87bd919dd97": {"model_name": "ButtonStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "button_color": "#dddd77", "font_weight": ""}}, "118bb3366cd448bfaeb0456dc885e8fa": {"model_name": "ButtonModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": ["medium"], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "ButtonView", "button_style": "", "description": "\ud83d\ude10", "disabled": false, "icon": "", "layout": "IPY_MODEL_d3238d0d8b78412a9f540b7ae05c5c08", "style": "IPY_MODEL_1c70e7abfdfd4d17a8cee87bd919dd97", "tooltip": "medium"}}, "92f71278234c46b4b74cf9eef2b20fee": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": "0.5em", "right": null, "top": null, "visibility": null, "width": "auto"}}, "5d994ac8ce2c41549ad6927707b2ed32": {"model_name": "ButtonStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "button_color": "#ffaaaa", "font_weight": ""}}, "2edc7e0442fc4c528e449d4a3fead894": {"model_name": "ButtonModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": ["sad"], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "ButtonView", "button_style": "", "description": "\ud83d\ude41", "disabled": false, "icon": "", "layout": "IPY_MODEL_92f71278234c46b4b74cf9eef2b20fee", "style": "IPY_MODEL_5d994ac8ce2c41549ad6927707b2ed32", "tooltip": "sad"}}, "3c23ca399fb64b5888f393c7709667f3": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": "auto"}}, "0f37bdb0136c44ca952326249ffc61a1": {"model_name": "DescriptionStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "DescriptionStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "description_width": ""}}, "f2ac3c2adea344f99f3858b743534152": {"model_name": "TextareaModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "TextareaModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "TextareaView", "continuous_update": true, "description": "", "description_tooltip": null, "disabled": false, "layout": "IPY_MODEL_3c23ca399fb64b5888f393c7709667f3", "placeholder": "We want your feedback!", "rows": null, "style": "IPY_MODEL_0f37bdb0136c44ca952326249ffc61a1", "value": ""}}, "da1e15d54d5445fb85abb543cda39458": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": "auto", "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": "auto"}}, "6d18307e1cfd4c26a4c23b90862e1367": {"model_name": "ButtonStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "button_color": null, "font_weight": ""}}, "7870582b8f72487d8b3d751e5a82739f": {"model_name": "ButtonModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "ButtonModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "ButtonView", "button_style": "", "description": "Submit", "disabled": false, "icon": "", "layout": "IPY_MODEL_da1e15d54d5445fb85abb543cda39458", "style": "IPY_MODEL_6d18307e1cfd4c26a4c23b90862e1367", "tooltip": ""}}, "36b4ddde17654ddc84ae1c003a818ec0": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": "none", "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "ee72efbdadbc404e8e645e902ce2b60a": {"model_name": "HBoxModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "HBoxModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "HBoxView", "box_style": "", "children": ["IPY_MODEL_f2ac3c2adea344f99f3858b743534152", "IPY_MODEL_7870582b8f72487d8b3d751e5a82739f"], "layout": "IPY_MODEL_36b4ddde17654ddc84ae1c003a818ec0"}}, "1d905f7079404a018476a616527cb314": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": "none", "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "d4ec01c3a28041cab19e19eb311cb2b5": {"model_name": "DescriptionStyleModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "DescriptionStyleModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "StyleView", "description_width": ""}}, "1cbde90c7cef45b8a017e3a24f4e24a8": {"model_name": "LabelModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "LabelModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "LabelView", "description": "", "description_tooltip": null, "layout": "IPY_MODEL_1d905f7079404a018476a616527cb314", "placeholder": "\u200b", "style": "IPY_MODEL_d4ec01c3a28041cab19e19eb311cb2b5", "value": "Thanks for your feedback!"}}, "c85f922050144fe4b916e95a58bdcc8c": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "9370886f5f5e4b96a87b3d91350ec983": {"model_name": "HBoxModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "HBoxModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "HBoxView", "box_style": "", "children": ["IPY_MODEL_23516cddded64f60ad1e8eb82fc32585", "IPY_MODEL_118bb3366cd448bfaeb0456dc885e8fa", "IPY_MODEL_2edc7e0442fc4c528e449d4a3fead894"], "layout": "IPY_MODEL_c85f922050144fe4b916e95a58bdcc8c"}}, "5ebcb6e394cc4094af1d2ac350cad942": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "4031938093104703adbbb21219a9d06a": {"model_name": "VBoxModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "VBoxModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "VBoxView", "box_style": "", "children": ["IPY_MODEL_9370886f5f5e4b96a87b3d91350ec983", "IPY_MODEL_ee72efbdadbc404e8e645e902ce2b60a", "IPY_MODEL_1cbde90c7cef45b8a017e3a24f4e24a8"], "layout": "IPY_MODEL_5ebcb6e394cc4094af1d2ac350cad942"}}, "4afa088b0e6b4f0581cec5fba040f179": {"model_name": "LayoutModel", "model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "state": {"_model_module": "@jupyter-widgets/base", "_model_module_version": "1.2.0", "_model_name": "LayoutModel", "_view_count": null, "_view_module": "@jupyter-widgets/base", "_view_module_version": "1.2.0", "_view_name": "LayoutView", "align_content": null, "align_items": null, "align_self": null, "border": null, "bottom": null, "display": null, "flex": null, "flex_flow": null, "grid_area": null, "grid_auto_columns": null, "grid_auto_flow": null, "grid_auto_rows": null, "grid_column": null, "grid_gap": null, "grid_row": null, "grid_template_areas": null, "grid_template_columns": null, "grid_template_rows": null, "height": null, "justify_content": null, "justify_items": null, "left": null, "margin": null, "max_height": null, "max_width": null, "min_height": null, "min_width": null, "object_fit": null, "object_position": null, "order": null, "overflow": null, "overflow_x": null, "overflow_y": null, "padding": null, "right": null, "top": null, "visibility": null, "width": null}}, "0e56ba5701c3409b9e6e0cf6657fef96": {"model_name": "VBoxModel", "model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "state": {"_dom_classes": [], "_model_module": "@jupyter-widgets/controls", "_model_module_version": "1.5.0", "_model_name": "VBoxModel", "_view_count": null, "_view_module": "@jupyter-widgets/controls", "_view_module_version": "1.5.0", "_view_name": "VBoxView", "box_style": "", "children": ["IPY_MODEL_4031938093104703adbbb21219a9d06a"], "layout": "IPY_MODEL_4afa088b0e6b4f0581cec5fba040f179"}}}, "version_major": 2, "version_minor": 0}
</script>
<script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./sequences/P1C3_RealEvolution/student"
        },
        predefinedOutput: true
    }
    </script>
<script>kernelName = 'python3'</script>
</div>
</main>
<footer class="footer-article noprint">
<!-- Previous / next buttons -->
<div class="prev-next-area">
<a class="left-prev" href="P1C3_Sequence3.html" id="prev-link" title="previous page">
<i class="fas fa-angle-left"></i>
<div class="prev-next-info">
<p class="prev-next-subtitle">previous</p>
<p class="prev-next-title"><strong>1.3.3: Competition and Interaction Transform Optimization into Game Theory</strong></p>
</div>
</a>
<a class="right-next" href="../../P1C4_LifeIsRL/student/P1C4_Title.html" id="next-link" title="next page">
<div class="prev-next-info">
<p class="prev-next-subtitle">next</p>
<p class="prev-next-title">Chapter 1.4 Life is Reinforcement Learning</p>
</div>
<i class="fas fa-angle-right"></i>
</a>
</div>
</footer>
</div>
</div>
<div class="footer-content row">
<footer class="col footer"><p>
<div class="extra_footer">
<div>
<a href="http://creativecommons.org/licenses/by/4.0/"><img src="https://i.creativecommons.org/l/by/4.0/88x31.png"/></a>
<a href="https://opensource.org/licenses/BSD-3-Clause"><img src="https://camo.githubusercontent.com/9b9ea65d95c9ef878afa1987df65731d47681336/68747470733a2f2f696d672e736869656c64732e696f2f707970692f6c2f736561626f726e2e737667"/></a>
The contents of this repository are shared under the <a href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.
Software elements are additionally licensed under the <a href="https://opensource.org/licenses/BSD-3-Clause">BSD (3-Clause) License</a>.
</div>
</div>
</p>
</footer>
</div>
</div>
</div>
</div>
<!-- Scripts loaded after <body> so the DOM is not blocked -->
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>
</body>
</html>